{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Gosling Demo"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this notebook, I am going to do an example gosling demo. This will include:\n",
    "1. A basic gosling visualization of gene expression data\n",
    "2. Several tutorials from the 2022 ISMB demo\n",
    "3. A recreation of [this](https://gosling.js.org/?example=LINKING) visualization using the gos package"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# imports\n",
    "import gosling as gos"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1. Gene expression data (warmup)\n",
    "To start, I am making a basic visual of gene expression data using the BIGWIG ChIP-seq dataset in the gosling documentation. This is to test out different configurations and how they impact a visualization.\n",
    "\n",
    "Dataset: [BigWig ChIP-seq for H2AZ](https://s3.amazonaws.com/gosling-lang.org/data/4DNFIMPI5A9N.bw), from gosling plain datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<!DOCTYPE html>\n",
       "<html>\n",
       "<head>\n",
       "  <style>.error { color: red; }</style>\n",
       "</head>\n",
       "<body>\n",
       "  <div id=\"jupyter-gosling-e160e32ecaa044628671734c7ebac699\"></div>\n",
       "  <script type=\"module\">\n",
       "    import { importWithMap } from \"https://unpkg.com/dynamic-importmap@0.1.0\";\n",
       "    let importMap = {\"imports\": {\"react\": \"https://esm.sh/react@18\", \"react-dom\": \"https://esm.sh/react-dom@18\", \"pixi\": \"https://esm.sh/pixi.js@6?bundle\", \"higlass\": \"https://esm.sh/higlass@1.13?external=react,react-dom,pixi&bundle\", \"gosling.js\": \"https://esm.sh/gosling.js@1.0.2?external=react,react-dom,pixi,higlass&bundle\"}};\n",
       "    let gosling = await importWithMap(\"gosling.js\", importMap);\n",
       "    let el = document.getElementById('jupyter-gosling-e160e32ecaa044628671734c7ebac699');\n",
       "    let spec = {\"tracks\": [{\"data\": {\"type\": \"bigwig\", \"url\": \"https://4dn-open-data-public.s3.amazonaws.com/fourfront-webprod/wfoutput/a57e91de-47f9-4157-9462-50fc78f357e3/4DNFIMPI5A9N.bw\", \"value\": \"line\"}, \"mark\": \"area\", \"x\": {\"domain\": {\"chromosome\": \"chr2\"}, \"field\": \"position\", \"type\": \"genomic\"}, \"y\": {\"field\": \"line\", \"type\": \"quantitative\"}}]};\n",
       "    let opt = {\"padding\": 0, \"theme\": null};\n",
       "    gosling.embed(el, spec, opt).catch((err) => {\n",
       "      el.innerHTML = `    <div class=\"error\">\n",
       "        <p>JavaScript Error: ${error.message}</p>\n",
       "        <p>This usually means there's a typo in your Gosling specification. See the javascript console for the full traceback.</p>\n",
       "    </div>`;\n",
       "      throw error;\n",
       "    });\n",
       "  </script>\n",
       "</body>\n",
       "</html>"
      ],
      "text/plain": [
       "View({\n",
       "  tracks: [Track({\n",
       "    data: BigWigData({\n",
       "      type: 'bigwig',\n",
       "      url: 'https://4dn-open-data-public.s3.amazonaws.com/fourfront-webprod/wfoutput/a57e91de-47f9-4157-9462-50fc78f357e3/4DNFIMPI5A9N.bw',\n",
       "      value: 'line'\n",
       "    }),\n",
       "    mark: 'area',\n",
       "    x: X({\n",
       "      domain: DomainChr({\n",
       "        chromosome: 'chr2'\n",
       "      }),\n",
       "      field: 'position',\n",
       "      type: 'genomic'\n",
       "    }),\n",
       "    y: Y({\n",
       "      field: 'line',\n",
       "      type: 'quantitative'\n",
       "    })\n",
       "  })]\n",
       "})"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# create a new track\n",
    "vis = gos.Track(\n",
    "    data=gos.BigWigData(\n",
    "        'bigwig',\n",
    "        url='https://4dn-open-data-public.s3.amazonaws.com/fourfront-webprod/wfoutput/a57e91de-47f9-4157-9462-50fc78f357e3/4DNFIMPI5A9N.bw',\n",
    "        # setting this value allows you to update the value of y!\n",
    "        #column='position'\n",
    "        value='line'\n",
    "    ),\n",
    ").mark_area(\n",
    "    # mark_area: sets the output to an area graph\n",
    "    # other option that may apply here --> mark_line\n",
    ").encode(\n",
    "    # can use this format for x, or set x to the column setting from the data declaration\n",
    "    x=gos.X('position:G', domain=gos.DomainChr('chr2')),\n",
    "    # DomainChr: allows you to zoom on a specific chromosome\n",
    "    y='line:Q'\n",
    ").view()\n",
    "\n",
    "# can also save as html to view in browser!\n",
    "display(vis)\n",
    "vis.save('visuals/single-track/basic-gene-expr.json')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Tutorials\n",
    "In this section, I'm going to learn more about writing in-depth gosling specs via the ISMB tutorials. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Single Track\n",
    "Data: [BED file](https://raw.githubusercontent.com/sehilyi/gemini-datasets/master/data/UCSC.HG38.Human.CytoBandIdeogram.bed) with cytoband information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<!DOCTYPE html>\n",
       "<html>\n",
       "<head>\n",
       "  <style>.error { color: red; }</style>\n",
       "</head>\n",
       "<body>\n",
       "  <div id=\"jupyter-gosling-557bc48ffa19499ba8167d5cde03ea95\"></div>\n",
       "  <script type=\"module\">\n",
       "    import { importWithMap } from \"https://unpkg.com/dynamic-importmap@0.1.0\";\n",
       "    let importMap = {\"imports\": {\"react\": \"https://esm.sh/react@18\", \"react-dom\": \"https://esm.sh/react-dom@18\", \"pixi\": \"https://esm.sh/pixi.js@6?bundle\", \"higlass\": \"https://esm.sh/higlass@1.13?external=react,react-dom,pixi&bundle\", \"gosling.js\": \"https://esm.sh/gosling.js@1.0.2?external=react,react-dom,pixi,higlass&bundle\"}};\n",
       "    let gosling = await importWithMap(\"gosling.js\", importMap);\n",
       "    let el = document.getElementById('jupyter-gosling-557bc48ffa19499ba8167d5cde03ea95');\n",
       "    let spec = {\"tracks\": [{\"data\": {\"type\": \"csv\", \"url\": \"https://raw.githubusercontent.com/sehilyi/gemini-datasets/master/data/UCSC.HG38.Human.CytoBandIdeogram.bed\", \"separator\": \"\\t\", \"headerNames\": [\"chrom\", \"chromStart\", \"chromEnd\", \"name\", \"stain\"], \"chromosomeField\": \"chrom\", \"genomicFields\": [\"chromStart\", \"chromEnd\"]}, \"mark\": \"point\", \"color\": {\"value\": \"lightblue\"}, \"x\": {\"field\": \"chromStart\", \"type\": \"genomic\"}}]};\n",
       "    let opt = {\"padding\": 0, \"theme\": null};\n",
       "    gosling.embed(el, spec, opt).catch((err) => {\n",
       "      el.innerHTML = `    <div class=\"error\">\n",
       "        <p>JavaScript Error: ${error.message}</p>\n",
       "        <p>This usually means there's a typo in your Gosling specification. See the javascript console for the full traceback.</p>\n",
       "    </div>`;\n",
       "      throw error;\n",
       "    });\n",
       "  </script>\n",
       "</body>\n",
       "</html>"
      ],
      "text/plain": [
       "View({\n",
       "  tracks: [Track({\n",
       "    color: ColorValue({\n",
       "      value: 'lightblue'\n",
       "    }),\n",
       "    data: {'type': 'csv', 'url': 'https://raw.githubusercontent.com/sehilyi/gemini-datasets/master/data/UCSC.HG38.Human.CytoBandIdeogram.bed', 'separator': '\\t', 'headerNames': ['chrom', 'chromStart', 'chromEnd', 'name', 'stain'], 'chromosomeField': 'chrom', 'genomicFields': ['chromStart', 'chromEnd']},\n",
       "    mark: 'point',\n",
       "    x: X({\n",
       "      field: 'chromStart',\n",
       "      type: 'genomic'\n",
       "    })\n",
       "  })]\n",
       "})"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# import data into csv wrapper\n",
    "data=gos.csv(\n",
    "    url='https://raw.githubusercontent.com/sehilyi/gemini-datasets/master/data/UCSC.HG38.Human.CytoBandIdeogram.bed',\n",
    "    separator=\"\\t\",\n",
    "    # identify headers and genomic fields\n",
    "    headerNames=['chrom', 'chromStart', 'chromEnd', 'name', 'stain'],\n",
    "    chromosomeField=\"chrom\",\n",
    "    genomicFields=[\"chromStart\", \"chromEnd\"]\n",
    ")\n",
    "\n",
    "# add data to track and view\n",
    "view = gos.Track(data).mark_point().encode(\n",
    "    # apply visual encoding\n",
    "    x=gos.X('chromStart', type='genomic'),\n",
    "    # or: x=gos.X('chromStart:G')\n",
    "    # including the Y code below will add a y axis with same data\n",
    "    #y=gos.Y('chromStart', type='genomic',axis='left'),\n",
    "    color=gos.value('lightblue')\n",
    ").view()\n",
    "\n",
    "# save as json and html (html for easy visual purposes)\n",
    "view.save('visuals/single-track/tutorial-intro.json')\n",
    "display(view)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<!DOCTYPE html>\n",
       "<html>\n",
       "<head>\n",
       "  <style>.error { color: red; }</style>\n",
       "</head>\n",
       "<body>\n",
       "  <div id=\"jupyter-gosling-3659fc4561ef42a599b1edbd72b80150\"></div>\n",
       "  <script type=\"module\">\n",
       "    import { importWithMap } from \"https://unpkg.com/dynamic-importmap@0.1.0\";\n",
       "    let importMap = {\"imports\": {\"react\": \"https://esm.sh/react@18\", \"react-dom\": \"https://esm.sh/react-dom@18\", \"pixi\": \"https://esm.sh/pixi.js@6?bundle\", \"higlass\": \"https://esm.sh/higlass@1.13?external=react,react-dom,pixi&bundle\", \"gosling.js\": \"https://esm.sh/gosling.js@1.0.2?external=react,react-dom,pixi,higlass&bundle\"}};\n",
       "    let gosling = await importWithMap(\"gosling.js\", importMap);\n",
       "    let el = document.getElementById('jupyter-gosling-3659fc4561ef42a599b1edbd72b80150');\n",
       "    let spec = {\"tracks\": [{\"data\": {\"type\": \"csv\", \"url\": \"https://raw.githubusercontent.com/sehilyi/gemini-datasets/master/data/UCSC.HG38.Human.CytoBandIdeogram.bed\", \"separator\": \"\\t\", \"headerNames\": [\"chrom\", \"chromStart\", \"chromEnd\", \"name\", \"stain\"], \"chromosomeField\": \"chrom\", \"genomicFields\": [\"chromStart\", \"chromEnd\"]}, \"mark\": \"triangleRight\", \"height\": 500, \"size\": {\"value\": 10}, \"width\": 500, \"x\": {\"field\": \"chromStart\", \"type\": \"genomic\"}, \"y\": {\"axis\": \"left\", \"field\": \"chromEnd\", \"type\": \"genomic\"}}]};\n",
       "    let opt = {\"padding\": 0, \"theme\": null};\n",
       "    gosling.embed(el, spec, opt).catch((err) => {\n",
       "      el.innerHTML = `    <div class=\"error\">\n",
       "        <p>JavaScript Error: ${error.message}</p>\n",
       "        <p>This usually means there's a typo in your Gosling specification. See the javascript console for the full traceback.</p>\n",
       "    </div>`;\n",
       "      throw error;\n",
       "    });\n",
       "  </script>\n",
       "</body>\n",
       "</html>"
      ],
      "text/plain": [
       "View({\n",
       "  tracks: [Track({\n",
       "    data: {'type': 'csv', 'url': 'https://raw.githubusercontent.com/sehilyi/gemini-datasets/master/data/UCSC.HG38.Human.CytoBandIdeogram.bed', 'separator': '\\t', 'headerNames': ['chrom', 'chromStart', 'chromEnd', 'name', 'stain'], 'chromosomeField': 'chrom', 'genomicFields': ['chromStart', 'chromEnd']},\n",
       "    height: 500,\n",
       "    mark: 'triangleRight',\n",
       "    size: SizeValue({\n",
       "      value: 10\n",
       "    }),\n",
       "    width: 500,\n",
       "    x: X({\n",
       "      field: 'chromStart',\n",
       "      type: 'genomic'\n",
       "    }),\n",
       "    y: Y({\n",
       "      axis: 'left',\n",
       "      field: 'chromEnd',\n",
       "      type: 'genomic'\n",
       "    })\n",
       "  })]\n",
       "})"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "'''\n",
    "exercise: add following changes to code:\n",
    "- add a size encoding with the constant value 10\n",
    "- change the mark type from point to triangleRight\n",
    "- add a y encoding to use the \"chromEnd\" field instead of \"chromStart\"\n",
    "- change the width and height of the track to be 500\n",
    "'''\n",
    "tracky = gos.Track(data).mark_triangleRight().encode(\n",
    "    x=gos.X(\"chromStart\", type=\"genomic\"),\n",
    "    y=gos.Y(\"chromEnd\", type=\"genomic\", axis=\"left\"),\n",
    "    # additional encodings ...\n",
    "    size=gos.value(10)\n",
    "\n",
    ").properties(\n",
    "    # track property overrides ...\n",
    "    width=500,\n",
    "    height=500\n",
    ").view()\n",
    "\n",
    "# view as html!\n",
    "tracky.save('visuals/single-track/tutorial-exercise.json')\n",
    "display(tracky)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we will expand this single-track visual by adding multiple domains and ranges, allowing for us to stratify each domain by color"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<!DOCTYPE html>\n",
       "<html>\n",
       "<head>\n",
       "  <style>.error { color: red; }</style>\n",
       "</head>\n",
       "<body>\n",
       "  <div id=\"jupyter-gosling-826830b564a34152b41c045e99f438a5\"></div>\n",
       "  <script type=\"module\">\n",
       "    import { importWithMap } from \"https://unpkg.com/dynamic-importmap@0.1.0\";\n",
       "    let importMap = {\"imports\": {\"react\": \"https://esm.sh/react@18\", \"react-dom\": \"https://esm.sh/react-dom@18\", \"pixi\": \"https://esm.sh/pixi.js@6?bundle\", \"higlass\": \"https://esm.sh/higlass@1.13?external=react,react-dom,pixi&bundle\", \"gosling.js\": \"https://esm.sh/gosling.js@1.0.2?external=react,react-dom,pixi,higlass&bundle\"}};\n",
       "    let gosling = await importWithMap(\"gosling.js\", importMap);\n",
       "    let el = document.getElementById('jupyter-gosling-826830b564a34152b41c045e99f438a5');\n",
       "    let spec = {\"tracks\": [{\"data\": {\"type\": \"csv\", \"url\": \"https://raw.githubusercontent.com/sehilyi/gemini-datasets/master/data/UCSC.HG38.Human.CytoBandIdeogram.bed\", \"separator\": \"\\t\", \"headerNames\": [\"chrom\", \"chromStart\", \"chromEnd\", \"name\", \"stain\"], \"chromosomeField\": \"chrom\", \"genomicFields\": [\"chromStart\", \"chromEnd\"]}, \"mark\": \"point\", \"color\": {\"domain\": [\"gneg\", \"gpos25\", \"gpos50\", \"gpos75\", \"gpos100\", \"gvar\"], \"field\": \"stain\", \"legend\": true, \"range\": [\"#265653\", \"#2A9D8F\", \"#8AB17D\", \"#E9C46A\", \"#F4A261\", \"#E76F51\"], \"type\": \"nominal\"}, \"x\": {\"field\": \"chromStart\", \"type\": \"genomic\"}, \"y\": {\"field\": \"stain\", \"type\": \"nominal\"}}]};\n",
       "    let opt = {\"padding\": 0, \"theme\": null};\n",
       "    gosling.embed(el, spec, opt).catch((err) => {\n",
       "      el.innerHTML = `    <div class=\"error\">\n",
       "        <p>JavaScript Error: ${error.message}</p>\n",
       "        <p>This usually means there's a typo in your Gosling specification. See the javascript console for the full traceback.</p>\n",
       "    </div>`;\n",
       "      throw error;\n",
       "    });\n",
       "  </script>\n",
       "</body>\n",
       "</html>"
      ],
      "text/plain": [
       "View({\n",
       "  tracks: [Track({\n",
       "    color: Color({\n",
       "      domain: ['gneg', 'gpos25', 'gpos50', 'gpos75', 'gpos100', 'gvar'],\n",
       "      field: 'stain',\n",
       "      legend: True,\n",
       "      range: ['#265653', '#2A9D8F', '#8AB17D', '#E9C46A', '#F4A261', '#E76F51'],\n",
       "      type: 'nominal'\n",
       "    }),\n",
       "    data: {'type': 'csv', 'url': 'https://raw.githubusercontent.com/sehilyi/gemini-datasets/master/data/UCSC.HG38.Human.CytoBandIdeogram.bed', 'separator': '\\t', 'headerNames': ['chrom', 'chromStart', 'chromEnd', 'name', 'stain'], 'chromosomeField': 'chrom', 'genomicFields': ['chromStart', 'chromEnd']},\n",
       "    mark: 'point',\n",
       "    x: X({\n",
       "      field: 'chromStart',\n",
       "      type: 'genomic'\n",
       "    }),\n",
       "    y: Y({\n",
       "      field: 'stain',\n",
       "      type: 'nominal'\n",
       "    })\n",
       "  })]\n",
       "})"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "color_strat= gos.Track(data).mark_point().encode(\n",
    "    x=gos.X('chromStart:G'),\n",
    "    y=gos.Y('stain:N'),\n",
    "    color=gos.Color(\n",
    "        # separate by color\n",
    "        'stain:N',\n",
    "        domain=[\"gneg\", \"gpos25\", \"gpos50\", \"gpos75\", \"gpos100\", \"gvar\"],\n",
    "        # define color per domain (recall mapping from domain to range!)\n",
    "        range=['#265653', '#2A9D8F', '#8AB17D', '#E9C46A', '#F4A261', '#E76F51'],\n",
    "        # add legend\n",
    "        legend=True\n",
    "    )\n",
    ").view()\n",
    "color_strat.save('visuals/single-track/stratified-color.json')\n",
    "display(color_strat)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Combining Tracks\n",
    "Now that we can do simple tracks, we can move on to composite tracks. This will use four subsets of a scATAC-seq dataset from Corces et al (2020)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<!DOCTYPE html>\n",
       "<html>\n",
       "<head>\n",
       "  <style>.error { color: red; }</style>\n",
       "</head>\n",
       "<body>\n",
       "  <div id=\"jupyter-gosling-900be8f1083d451a97e451dad401b6cd\"></div>\n",
       "  <script type=\"module\">\n",
       "    import { importWithMap } from \"https://unpkg.com/dynamic-importmap@0.1.0\";\n",
       "    let importMap = {\"imports\": {\"react\": \"https://esm.sh/react@18\", \"react-dom\": \"https://esm.sh/react-dom@18\", \"pixi\": \"https://esm.sh/pixi.js@6?bundle\", \"higlass\": \"https://esm.sh/higlass@1.13?external=react,react-dom,pixi&bundle\", \"gosling.js\": \"https://esm.sh/gosling.js@1.0.2?external=react,react-dom,pixi,higlass&bundle\"}};\n",
       "    let gosling = await importWithMap(\"gosling.js\", importMap);\n",
       "    let el = document.getElementById('jupyter-gosling-900be8f1083d451a97e451dad401b6cd');\n",
       "    let spec = {\"tracks\": [{\"data\": {\"type\": \"bigwig\", \"url\": \"https://s3.amazonaws.com/gosling-lang.org/data/ExcitatoryNeurons-insertions_bin100_RIPnorm.bw\", \"column\": \"position\", \"value\": \"peak\"}, \"mark\": \"line\", \"color\": {\"value\": \"black\"}, \"height\": 100, \"title\": \"\", \"x\": {\"field\": \"position\", \"type\": \"genomic\"}, \"y\": {\"field\": \"peak\", \"type\": \"quantitative\"}}], \"title\": \"Line view\"};\n",
       "    let opt = {\"padding\": 0, \"theme\": null};\n",
       "    gosling.embed(el, spec, opt).catch((err) => {\n",
       "      el.innerHTML = `    <div class=\"error\">\n",
       "        <p>JavaScript Error: ${error.message}</p>\n",
       "        <p>This usually means there's a typo in your Gosling specification. See the javascript console for the full traceback.</p>\n",
       "    </div>`;\n",
       "      throw error;\n",
       "    });\n",
       "  </script>\n",
       "</body>\n",
       "</html>"
      ],
      "text/plain": [
       "View({\n",
       "  title: 'Line view',\n",
       "  tracks: [Track({\n",
       "    color: ColorValue({\n",
       "      value: 'black'\n",
       "    }),\n",
       "    data: {'type': 'bigwig', 'url': 'https://s3.amazonaws.com/gosling-lang.org/data/ExcitatoryNeurons-insertions_bin100_RIPnorm.bw', 'column': 'position', 'value': 'peak'},\n",
       "    height: 100,\n",
       "    mark: 'line',\n",
       "    title: '',\n",
       "    x: X({\n",
       "      field: 'position',\n",
       "      type: 'genomic'\n",
       "    }),\n",
       "    y: Y({\n",
       "      field: 'peak',\n",
       "      type: 'quantitative'\n",
       "    })\n",
       "  })]\n",
       "})"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "\n",
       "<!DOCTYPE html>\n",
       "<html>\n",
       "<head>\n",
       "  <style>.error { color: red; }</style>\n",
       "</head>\n",
       "<body>\n",
       "  <div id=\"jupyter-gosling-5886a5f91fed436e86ef26e6ec6d2960\"></div>\n",
       "  <script type=\"module\">\n",
       "    import { importWithMap } from \"https://unpkg.com/dynamic-importmap@0.1.0\";\n",
       "    let importMap = {\"imports\": {\"react\": \"https://esm.sh/react@18\", \"react-dom\": \"https://esm.sh/react-dom@18\", \"pixi\": \"https://esm.sh/pixi.js@6?bundle\", \"higlass\": \"https://esm.sh/higlass@1.13?external=react,react-dom,pixi&bundle\", \"gosling.js\": \"https://esm.sh/gosling.js@1.0.2?external=react,react-dom,pixi,higlass&bundle\"}};\n",
       "    let gosling = await importWithMap(\"gosling.js\", importMap);\n",
       "    let el = document.getElementById('jupyter-gosling-5886a5f91fed436e86ef26e6ec6d2960');\n",
       "    let spec = {\"tracks\": [{\"data\": {\"type\": \"bigwig\", \"url\": \"https://s3.amazonaws.com/gosling-lang.org/data/ExcitatoryNeurons-insertions_bin100_RIPnorm.bw\", \"column\": \"position\", \"value\": \"peak\"}, \"mark\": \"point\", \"color\": {\"field\": \"peak\", \"type\": \"quantitative\"}, \"height\": 100, \"title\": \"\", \"x\": {\"field\": \"position\", \"type\": \"genomic\"}, \"y\": {\"field\": \"peak\", \"type\": \"quantitative\"}}], \"title\": \"Colored point view\"};\n",
       "    let opt = {\"padding\": 0, \"theme\": null};\n",
       "    gosling.embed(el, spec, opt).catch((err) => {\n",
       "      el.innerHTML = `    <div class=\"error\">\n",
       "        <p>JavaScript Error: ${error.message}</p>\n",
       "        <p>This usually means there's a typo in your Gosling specification. See the javascript console for the full traceback.</p>\n",
       "    </div>`;\n",
       "      throw error;\n",
       "    });\n",
       "  </script>\n",
       "</body>\n",
       "</html>"
      ],
      "text/plain": [
       "View({\n",
       "  title: 'Colored point view',\n",
       "  tracks: [Track({\n",
       "    color: Color({\n",
       "      field: 'peak',\n",
       "      type: 'quantitative'\n",
       "    }),\n",
       "    data: {'type': 'bigwig', 'url': 'https://s3.amazonaws.com/gosling-lang.org/data/ExcitatoryNeurons-insertions_bin100_RIPnorm.bw', 'column': 'position', 'value': 'peak'},\n",
       "    height: 100,\n",
       "    mark: 'point',\n",
       "    title: '',\n",
       "    x: X({\n",
       "      field: 'position',\n",
       "      type: 'genomic'\n",
       "    }),\n",
       "    y: Y({\n",
       "      field: 'peak',\n",
       "      type: 'quantitative'\n",
       "    })\n",
       "  })]\n",
       "})"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "\n",
       "<!DOCTYPE html>\n",
       "<html>\n",
       "<head>\n",
       "  <style>.error { color: red; }</style>\n",
       "</head>\n",
       "<body>\n",
       "  <div id=\"jupyter-gosling-71b187d7042048ddb5677a6ca8342605\"></div>\n",
       "  <script type=\"module\">\n",
       "    import { importWithMap } from \"https://unpkg.com/dynamic-importmap@0.1.0\";\n",
       "    let importMap = {\"imports\": {\"react\": \"https://esm.sh/react@18\", \"react-dom\": \"https://esm.sh/react-dom@18\", \"pixi\": \"https://esm.sh/pixi.js@6?bundle\", \"higlass\": \"https://esm.sh/higlass@1.13?external=react,react-dom,pixi&bundle\", \"gosling.js\": \"https://esm.sh/gosling.js@1.0.2?external=react,react-dom,pixi,higlass&bundle\"}};\n",
       "    let gosling = await importWithMap(\"gosling.js\", importMap);\n",
       "    let el = document.getElementById('jupyter-gosling-71b187d7042048ddb5677a6ca8342605');\n",
       "    let spec = {\"alignment\": \"overlay\", \"tracks\": [{\"data\": {\"type\": \"bigwig\", \"url\": \"https://s3.amazonaws.com/gosling-lang.org/data/ExcitatoryNeurons-insertions_bin100_RIPnorm.bw\", \"column\": \"position\", \"value\": \"peak\"}, \"mark\": \"line\", \"color\": {\"value\": \"black\"}, \"height\": 100, \"title\": \"\", \"x\": {\"field\": \"position\", \"type\": \"genomic\"}, \"y\": {\"field\": \"peak\", \"type\": \"quantitative\"}}, {\"data\": {\"type\": \"bigwig\", \"url\": \"https://s3.amazonaws.com/gosling-lang.org/data/ExcitatoryNeurons-insertions_bin100_RIPnorm.bw\", \"column\": \"position\", \"value\": \"peak\"}, \"mark\": \"point\", \"color\": {\"field\": \"peak\", \"type\": \"quantitative\"}, \"height\": 100, \"title\": \"\", \"x\": {\"field\": \"position\", \"type\": \"genomic\"}, \"y\": {\"field\": \"peak\", \"type\": \"quantitative\"}}], \"title\": \"Overlaid view\"};\n",
       "    let opt = {\"padding\": 0, \"theme\": null};\n",
       "    gosling.embed(el, spec, opt).catch((err) => {\n",
       "      el.innerHTML = `    <div class=\"error\">\n",
       "        <p>JavaScript Error: ${error.message}</p>\n",
       "        <p>This usually means there's a typo in your Gosling specification. See the javascript console for the full traceback.</p>\n",
       "    </div>`;\n",
       "      throw error;\n",
       "    });\n",
       "  </script>\n",
       "</body>\n",
       "</html>"
      ],
      "text/plain": [
       "View({\n",
       "  alignment: 'overlay',\n",
       "  title: 'Overlaid view',\n",
       "  tracks: (Track({\n",
       "    color: ColorValue({\n",
       "      value: 'black'\n",
       "    }),\n",
       "    data: {'type': 'bigwig', 'url': 'https://s3.amazonaws.com/gosling-lang.org/data/ExcitatoryNeurons-insertions_bin100_RIPnorm.bw', 'column': 'position', 'value': 'peak'},\n",
       "    height: 100,\n",
       "    mark: 'line',\n",
       "    title: '',\n",
       "    x: X({\n",
       "      field: 'position',\n",
       "      type: 'genomic'\n",
       "    }),\n",
       "    y: Y({\n",
       "      field: 'peak',\n",
       "      type: 'quantitative'\n",
       "    })\n",
       "  }), Track({\n",
       "    color: Color({\n",
       "      field: 'peak',\n",
       "      type: 'quantitative'\n",
       "    }),\n",
       "    data: {'type': 'bigwig', 'url': 'https://s3.amazonaws.com/gosling-lang.org/data/ExcitatoryNeurons-insertions_bin100_RIPnorm.bw', 'column': 'position', 'value': 'peak'},\n",
       "    height: 100,\n",
       "    mark: 'point',\n",
       "    title: '',\n",
       "    x: X({\n",
       "      field: 'position',\n",
       "      type: 'genomic'\n",
       "    }),\n",
       "    y: Y({\n",
       "      field: 'peak',\n",
       "      type: 'quantitative'\n",
       "    })\n",
       "  }))\n",
       "})"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# access data through url list\n",
    "urls = [\n",
    "    f\"https://s3.amazonaws.com/gosling-lang.org/data/{file}\"\n",
    "    for file in [\n",
    "        \"ExcitatoryNeurons-insertions_bin100_RIPnorm.bw\",\n",
    "        \"InhibitoryNeurons-insertions_bin100_RIPnorm.bw\",\n",
    "        \"Microglia-insertions_bin100_RIPnorm.bw\",\n",
    "        \"Astrocytes-insertions_bin100_RIPnorm.bw\",\n",
    "    ]\n",
    "]\n",
    "\n",
    "# use data from first url\n",
    "data = gos.bigwig(\n",
    "    urls[0], \n",
    "    column='position', \n",
    "    value='peak'\n",
    ")\n",
    "\n",
    "# use base track to derive other tracks --> do not have to repeat defs\n",
    "base = gos.Track(data).encode(\n",
    "    x=gos.X('position:G')\n",
    ").properties(height=100)\n",
    "\n",
    "heatmap = base.mark_rect().encode(\n",
    "    # color code based on the size of the peaks to create a heatmap\n",
    "    color=gos.Color('peak:Q')\n",
    ").view()\n",
    "\n",
    "line_graph = base.mark_line().encode(\n",
    "    # create line graph\n",
    "    y=gos.Y(\"peak:Q\"),\n",
    "    color=gos.value(\"black\")\n",
    ").properties(title=\"\")\n",
    "display(line_graph.view().properties(title=\"Line view\"))\n",
    "\n",
    "points = line_graph.mark_point()\n",
    "\n",
    "colored_points = points.encode(\n",
    "    # add color to points corresponding w/ height\n",
    "    color=gos.Color('peak:Q'),\n",
    ").properties(title=\"\")\n",
    "display(colored_points.view().properties(title=\"Colored point view\"))\n",
    "\n",
    "# overlay the two views.\n",
    "overlaid_view = gos.overlay(line_graph, colored_points)\n",
    "\n",
    "overlaid_view.save('visuals/single-track/single_track_overlay.json')\n",
    "display(overlaid_view.properties(title=\"Overlaid view\"))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we can reuse tracks to change our visual and add overlays, we can now use the same track definition to create visuals of other data sources!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<!DOCTYPE html>\n",
       "<html>\n",
       "<head>\n",
       "  <style>.error { color: red; }</style>\n",
       "</head>\n",
       "<body>\n",
       "  <div id=\"jupyter-gosling-86bd803656884cd4a1e7958e9d6dda05\"></div>\n",
       "  <script type=\"module\">\n",
       "    import { importWithMap } from \"https://unpkg.com/dynamic-importmap@0.1.0\";\n",
       "    let importMap = {\"imports\": {\"react\": \"https://esm.sh/react@18\", \"react-dom\": \"https://esm.sh/react-dom@18\", \"pixi\": \"https://esm.sh/pixi.js@6?bundle\", \"higlass\": \"https://esm.sh/higlass@1.13?external=react,react-dom,pixi&bundle\", \"gosling.js\": \"https://esm.sh/gosling.js@1.0.2?external=react,react-dom,pixi,higlass&bundle\"}};\n",
       "    let gosling = await importWithMap(\"gosling.js\", importMap);\n",
       "    let el = document.getElementById('jupyter-gosling-86bd803656884cd4a1e7958e9d6dda05');\n",
       "    let spec = {\"tracks\": [{\"data\": {\"type\": \"bigwig\", \"url\": \"https://s3.amazonaws.com/gosling-lang.org/data/ExcitatoryNeurons-insertions_bin100_RIPnorm.bw\", \"column\": \"position\", \"value\": \"peak\"}, \"mark\": \"bar\", \"height\": 80, \"title\": \"Iterative view of plot 0\", \"x\": {\"field\": \"position\", \"type\": \"genomic\"}, \"y\": {\"axis\": \"right\", \"field\": \"peak\", \"type\": \"quantitative\"}}]};\n",
       "    let opt = {\"padding\": 0, \"theme\": null};\n",
       "    gosling.embed(el, spec, opt).catch((err) => {\n",
       "      el.innerHTML = `    <div class=\"error\">\n",
       "        <p>JavaScript Error: ${error.message}</p>\n",
       "        <p>This usually means there's a typo in your Gosling specification. See the javascript console for the full traceback.</p>\n",
       "    </div>`;\n",
       "      throw error;\n",
       "    });\n",
       "  </script>\n",
       "</body>\n",
       "</html>"
      ],
      "text/plain": [
       "View({\n",
       "  tracks: [Track({\n",
       "    data: {'type': 'bigwig', 'url': 'https://s3.amazonaws.com/gosling-lang.org/data/ExcitatoryNeurons-insertions_bin100_RIPnorm.bw', 'column': 'position', 'value': 'peak'},\n",
       "    height: 80,\n",
       "    mark: 'bar',\n",
       "    title: 'Iterative view of plot 0',\n",
       "    x: X({\n",
       "      field: 'position',\n",
       "      type: 'genomic'\n",
       "    }),\n",
       "    y: Y({\n",
       "      axis: 'right',\n",
       "      field: 'peak',\n",
       "      type: 'quantitative'\n",
       "    })\n",
       "  })]\n",
       "})"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "\n",
       "<!DOCTYPE html>\n",
       "<html>\n",
       "<head>\n",
       "  <style>.error { color: red; }</style>\n",
       "</head>\n",
       "<body>\n",
       "  <div id=\"jupyter-gosling-38b0b0b93de049af8922e33c389b4738\"></div>\n",
       "  <script type=\"module\">\n",
       "    import { importWithMap } from \"https://unpkg.com/dynamic-importmap@0.1.0\";\n",
       "    let importMap = {\"imports\": {\"react\": \"https://esm.sh/react@18\", \"react-dom\": \"https://esm.sh/react-dom@18\", \"pixi\": \"https://esm.sh/pixi.js@6?bundle\", \"higlass\": \"https://esm.sh/higlass@1.13?external=react,react-dom,pixi&bundle\", \"gosling.js\": \"https://esm.sh/gosling.js@1.0.2?external=react,react-dom,pixi,higlass&bundle\"}};\n",
       "    let gosling = await importWithMap(\"gosling.js\", importMap);\n",
       "    let el = document.getElementById('jupyter-gosling-38b0b0b93de049af8922e33c389b4738');\n",
       "    let spec = {\"tracks\": [{\"data\": {\"type\": \"bigwig\", \"url\": \"https://s3.amazonaws.com/gosling-lang.org/data/InhibitoryNeurons-insertions_bin100_RIPnorm.bw\", \"column\": \"position\", \"value\": \"peak\"}, \"mark\": \"bar\", \"height\": 80, \"title\": \"Iterative view of plot 1\", \"x\": {\"field\": \"position\", \"type\": \"genomic\"}, \"y\": {\"axis\": \"right\", \"field\": \"peak\", \"type\": \"quantitative\"}}]};\n",
       "    let opt = {\"padding\": 0, \"theme\": null};\n",
       "    gosling.embed(el, spec, opt).catch((err) => {\n",
       "      el.innerHTML = `    <div class=\"error\">\n",
       "        <p>JavaScript Error: ${error.message}</p>\n",
       "        <p>This usually means there's a typo in your Gosling specification. See the javascript console for the full traceback.</p>\n",
       "    </div>`;\n",
       "      throw error;\n",
       "    });\n",
       "  </script>\n",
       "</body>\n",
       "</html>"
      ],
      "text/plain": [
       "View({\n",
       "  tracks: [Track({\n",
       "    data: {'type': 'bigwig', 'url': 'https://s3.amazonaws.com/gosling-lang.org/data/InhibitoryNeurons-insertions_bin100_RIPnorm.bw', 'column': 'position', 'value': 'peak'},\n",
       "    height: 80,\n",
       "    mark: 'bar',\n",
       "    title: 'Iterative view of plot 1',\n",
       "    x: X({\n",
       "      field: 'position',\n",
       "      type: 'genomic'\n",
       "    }),\n",
       "    y: Y({\n",
       "      axis: 'right',\n",
       "      field: 'peak',\n",
       "      type: 'quantitative'\n",
       "    })\n",
       "  })]\n",
       "})"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "\n",
       "<!DOCTYPE html>\n",
       "<html>\n",
       "<head>\n",
       "  <style>.error { color: red; }</style>\n",
       "</head>\n",
       "<body>\n",
       "  <div id=\"jupyter-gosling-669d7261af04421885c1abdeb8dcd6eb\"></div>\n",
       "  <script type=\"module\">\n",
       "    import { importWithMap } from \"https://unpkg.com/dynamic-importmap@0.1.0\";\n",
       "    let importMap = {\"imports\": {\"react\": \"https://esm.sh/react@18\", \"react-dom\": \"https://esm.sh/react-dom@18\", \"pixi\": \"https://esm.sh/pixi.js@6?bundle\", \"higlass\": \"https://esm.sh/higlass@1.13?external=react,react-dom,pixi&bundle\", \"gosling.js\": \"https://esm.sh/gosling.js@1.0.2?external=react,react-dom,pixi,higlass&bundle\"}};\n",
       "    let gosling = await importWithMap(\"gosling.js\", importMap);\n",
       "    let el = document.getElementById('jupyter-gosling-669d7261af04421885c1abdeb8dcd6eb');\n",
       "    let spec = {\"tracks\": [{\"data\": {\"type\": \"bigwig\", \"url\": \"https://s3.amazonaws.com/gosling-lang.org/data/Microglia-insertions_bin100_RIPnorm.bw\", \"column\": \"position\", \"value\": \"peak\"}, \"mark\": \"bar\", \"height\": 80, \"title\": \"Iterative view of plot 2\", \"x\": {\"field\": \"position\", \"type\": \"genomic\"}, \"y\": {\"axis\": \"right\", \"field\": \"peak\", \"type\": \"quantitative\"}}]};\n",
       "    let opt = {\"padding\": 0, \"theme\": null};\n",
       "    gosling.embed(el, spec, opt).catch((err) => {\n",
       "      el.innerHTML = `    <div class=\"error\">\n",
       "        <p>JavaScript Error: ${error.message}</p>\n",
       "        <p>This usually means there's a typo in your Gosling specification. See the javascript console for the full traceback.</p>\n",
       "    </div>`;\n",
       "      throw error;\n",
       "    });\n",
       "  </script>\n",
       "</body>\n",
       "</html>"
      ],
      "text/plain": [
       "View({\n",
       "  tracks: [Track({\n",
       "    data: {'type': 'bigwig', 'url': 'https://s3.amazonaws.com/gosling-lang.org/data/Microglia-insertions_bin100_RIPnorm.bw', 'column': 'position', 'value': 'peak'},\n",
       "    height: 80,\n",
       "    mark: 'bar',\n",
       "    title: 'Iterative view of plot 2',\n",
       "    x: X({\n",
       "      field: 'position',\n",
       "      type: 'genomic'\n",
       "    }),\n",
       "    y: Y({\n",
       "      axis: 'right',\n",
       "      field: 'peak',\n",
       "      type: 'quantitative'\n",
       "    })\n",
       "  })]\n",
       "})"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "\n",
       "<!DOCTYPE html>\n",
       "<html>\n",
       "<head>\n",
       "  <style>.error { color: red; }</style>\n",
       "</head>\n",
       "<body>\n",
       "  <div id=\"jupyter-gosling-fd29f39d3c384c1fbde91ca7cfbc735d\"></div>\n",
       "  <script type=\"module\">\n",
       "    import { importWithMap } from \"https://unpkg.com/dynamic-importmap@0.1.0\";\n",
       "    let importMap = {\"imports\": {\"react\": \"https://esm.sh/react@18\", \"react-dom\": \"https://esm.sh/react-dom@18\", \"pixi\": \"https://esm.sh/pixi.js@6?bundle\", \"higlass\": \"https://esm.sh/higlass@1.13?external=react,react-dom,pixi&bundle\", \"gosling.js\": \"https://esm.sh/gosling.js@1.0.2?external=react,react-dom,pixi,higlass&bundle\"}};\n",
       "    let gosling = await importWithMap(\"gosling.js\", importMap);\n",
       "    let el = document.getElementById('jupyter-gosling-fd29f39d3c384c1fbde91ca7cfbc735d');\n",
       "    let spec = {\"tracks\": [{\"data\": {\"type\": \"bigwig\", \"url\": \"https://s3.amazonaws.com/gosling-lang.org/data/Astrocytes-insertions_bin100_RIPnorm.bw\", \"column\": \"position\", \"value\": \"peak\"}, \"mark\": \"bar\", \"height\": 80, \"title\": \"Iterative view of plot 3\", \"x\": {\"field\": \"position\", \"type\": \"genomic\"}, \"y\": {\"axis\": \"right\", \"field\": \"peak\", \"type\": \"quantitative\"}}]};\n",
       "    let opt = {\"padding\": 0, \"theme\": null};\n",
       "    gosling.embed(el, spec, opt).catch((err) => {\n",
       "      el.innerHTML = `    <div class=\"error\">\n",
       "        <p>JavaScript Error: ${error.message}</p>\n",
       "        <p>This usually means there's a typo in your Gosling specification. See the javascript console for the full traceback.</p>\n",
       "    </div>`;\n",
       "      throw error;\n",
       "    });\n",
       "  </script>\n",
       "</body>\n",
       "</html>"
      ],
      "text/plain": [
       "View({\n",
       "  tracks: [Track({\n",
       "    data: {'type': 'bigwig', 'url': 'https://s3.amazonaws.com/gosling-lang.org/data/Astrocytes-insertions_bin100_RIPnorm.bw', 'column': 'position', 'value': 'peak'},\n",
       "    height: 80,\n",
       "    mark: 'bar',\n",
       "    title: 'Iterative view of plot 3',\n",
       "    x: X({\n",
       "      field: 'position',\n",
       "      type: 'genomic'\n",
       "    }),\n",
       "    y: Y({\n",
       "      axis: 'right',\n",
       "      field: 'peak',\n",
       "      type: 'quantitative'\n",
       "    })\n",
       "  })]\n",
       "})"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# create function to generate barplot for any scATAC-seq\n",
    "def barplot(url: str, title: str=None, color: str=None):\n",
    "    data=gos.bigwig(\n",
    "        url=url, \n",
    "        column='position', \n",
    "        value='peak'\n",
    "    )\n",
    "    track=gos.Track(data).mark_bar().encode(\n",
    "        x=gos.X('position:G'),\n",
    "        y=gos.Y('peak:Q', axis='right')\n",
    "    )\n",
    "    if color:\n",
    "        track=track.encode(color=gos.value(color))\n",
    "    if title:\n",
    "        track=track.properties(title=title)\n",
    "    return track.properties(height=80)\n",
    "\n",
    "# iterate through and save each json!\n",
    "for i in range(0, len(urls)):\n",
    "    plot = barplot(urls[i], f'Iterative view of plot {i}').view()\n",
    "    plot.save(f'visuals/single-track/iterative-view-{i}.json')\n",
    "    display(plot)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Suppose we want to align these plots. We can zip the colors and tracks such that\n",
    "the colors will be divided based on url and the barplots can be aligned."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_tracks = []\n",
    "\n",
    "for url, color in zip(urls, ['#2A9D8F', '#8AB17D', '#E9C46A', '#F4A261']):\n",
    "    title = url.split(\"/\")[-1].split(\"-\")[0]\n",
    "    track = barplot(url=url, title=title, color=color)\n",
    "    all_tracks.append(track)\n",
    "\n",
    "# use gos.stack to stack into a view\n",
    "gos.stack(*all_tracks).properties(\n",
    "    xDomain=(gos.GenomicDomain(chromosome='1'))\n",
    ").save('visuals/single-track/stacked_views.json')\n",
    "\n",
    "# and, for good measure, a craaaazy mess of a visualization: overlay\n",
    "gos.overlay(*all_tracks).properties(\n",
    "    xDomain=(gos.GenomicDomain(chromosome='1')),\n",
    ").save('visuals/single-track/overlaid_bars.json')\n",
    "# what a terrible visualization. cool i could make it though :D"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3. Visual Linking\n",
    "The goal of this section is to recreate [this](https://gosling.js.org/?example=LINKING) visualization using gos\n",
    "\n",
    "How is this being structured?\n",
    "- view with 3 tracks\n",
    "    - one subview (horizontal) and one track\n",
    "\n",
    "To do this, we can structure from the individual tracks up to the views (recall that alignment happens in the view level!)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can complete this by writing each track individually, then combining them into a big view. How do you eat an elephant? You chunk it down."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# first, define the data\n",
    "data=gos.MultivecData(\n",
    "        type='multivec',\n",
    "        url='https://server.gosling-lang.org/api/v1/tileset_info/?d=cistrome-multivec',\n",
    "        row='sample',\n",
    "        column='position',\n",
    "        value='peak',\n",
    "        # define categories for use in individual overlays!\n",
    "        categories = ['sample 1', 'sample 2', 'sample 3', 'sample 4'],\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create upper left track\n",
    "base_chart=gos.Track(\n",
    "    data=data,\n",
    "    ).mark_bar().encode(\n",
    "        # use start and end to define x\n",
    "        row='sample:N',\n",
    "        x=gos.X('start:G', \n",
    "                domain=gos.GenomicDomain(gos.DomainChr(chromosome='chr1')),),\n",
    "        xe='end:G',\n",
    "        y='peak:Q',\n",
    "        color=gos.Color('sample:N'),\n",
    "    )\n",
    "\n",
    "# view chart\n",
    "#base_chart.view().save('visuals/recreation/base_view.json')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will use this base chart as both our linear and circular visualsâ€” both will have the brush functionality sweeping over the chart. Advantage of using a track is that you can make two views with it!!\n",
    "\n",
    "Now, we are going to make the brush and detailed view. The detailed view will be based on our initial base_chart, but will have a smaller genomic domain."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "detailed_track=base_chart.encode(\n",
    "    x=gos.X(\n",
    "        'start:G',\n",
    "        # use style to set background color + opacity\n",
    "        domain=gos.GenomicDomain(gos.DomainChrInterval(chromosome='chr1', interval=[160000000, 200000000])),\n",
    "        # use linking id\n",
    "        linkingId='brushId'\n",
    "    ),\n",
    "    style=gos.Style(background='steelblue', backgroundOpacity=0.2),\n",
    ").properties(\n",
    "    width=690,\n",
    "    height=200,\n",
    "    title=\"Detailed View\"\n",
    ")\n",
    "\n",
    "# view chart\n",
    "#detailed_track.view().save('visuals/recreation/detail_view.json')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we can make the brush track views. This will entail \n",
    "1. Making the circle_view from the base view and adding a brush\n",
    "2. Making the linear_view from the base view and adding a brush"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "brush_track=base_chart.mark_brush().encode(\n",
    "    x=gos.X(\n",
    "        'start:G',\n",
    "        linkingId='brushId'\n",
    "    ),\n",
    "    color=gos.value('steelblue')\n",
    ")\n",
    "\n",
    "# create views\n",
    "circle_brush_track=base_chart.properties(\n",
    "    layout='circular',\n",
    "    static=True,\n",
    ")\n",
    "\n",
    "linear_brush_track=base_chart.properties(\n",
    "    layout='linear',\n",
    "    static=True,\n",
    ")\n",
    "\n",
    "# overlay brush onto views\n",
    "circ_overview=gos.overlay(\n",
    "    circle_brush_track,\n",
    "    brush_track\n",
    ").properties(\n",
    "    layout='circular',\n",
    "    centerRadius=0.4,\n",
    "    width=250,\n",
    "    height=130,\n",
    ")\n",
    "\n",
    "lin_overview=gos.overlay(\n",
    "    linear_brush_track,\n",
    "    brush_track\n",
    ").properties(\n",
    "    height=200,\n",
    "    width=400,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, create a composition of the sections\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<!DOCTYPE html>\n",
       "<html>\n",
       "<head>\n",
       "  <style>.error { color: red; }</style>\n",
       "</head>\n",
       "<body>\n",
       "  <div id=\"jupyter-gosling-d74880836d454983862980dc04b592d1\"></div>\n",
       "  <script type=\"module\">\n",
       "    import { importWithMap } from \"https://unpkg.com/dynamic-importmap@0.1.0\";\n",
       "    let importMap = {\"imports\": {\"react\": \"https://esm.sh/react@18\", \"react-dom\": \"https://esm.sh/react-dom@18\", \"pixi\": \"https://esm.sh/pixi.js@6?bundle\", \"higlass\": \"https://esm.sh/higlass@1.13?external=react,react-dom,pixi&bundle\", \"gosling.js\": \"https://esm.sh/gosling.js@1.0.2?external=react,react-dom,pixi,higlass&bundle\"}};\n",
       "    let gosling = await importWithMap(\"gosling.js\", importMap);\n",
       "    let el = document.getElementById('jupyter-gosling-d74880836d454983862980dc04b592d1');\n",
       "    let spec = {\"arrangement\": \"vertical\", \"views\": [{\"arrangement\": \"horizontal\", \"views\": [{\"alignment\": \"overlay\", \"tracks\": [{\"data\": {\"type\": \"multivec\", \"url\": \"https://server.gosling-lang.org/api/v1/tileset_info/?d=cistrome-multivec\", \"categories\": [\"sample 1\", \"sample 2\", \"sample 3\", \"sample 4\"], \"column\": \"position\", \"row\": \"sample\", \"value\": \"peak\"}, \"mark\": \"bar\", \"color\": {\"field\": \"sample\", \"type\": \"nominal\"}, \"layout\": \"circular\", \"row\": {\"field\": \"sample\", \"type\": \"nominal\"}, \"static\": true, \"x\": {\"domain\": {\"chromosome\": \"chr1\"}, \"field\": \"start\", \"type\": \"genomic\"}, \"xe\": {\"field\": \"end\", \"type\": \"genomic\"}, \"y\": {\"field\": \"peak\", \"type\": \"quantitative\"}}, {\"data\": {\"type\": \"multivec\", \"url\": \"https://server.gosling-lang.org/api/v1/tileset_info/?d=cistrome-multivec\", \"categories\": [\"sample 1\", \"sample 2\", \"sample 3\", \"sample 4\"], \"column\": \"position\", \"row\": \"sample\", \"value\": \"peak\"}, \"mark\": \"brush\", \"color\": {\"value\": \"steelblue\"}, \"row\": {\"field\": \"sample\", \"type\": \"nominal\"}, \"x\": {\"field\": \"start\", \"linkingId\": \"brushId\", \"type\": \"genomic\"}, \"xe\": {\"field\": \"end\", \"type\": \"genomic\"}, \"y\": {\"field\": \"peak\", \"type\": \"quantitative\"}}], \"layout\": \"circular\", \"centerRadius\": 0.4, \"width\": 250, \"height\": 130}, {\"alignment\": \"overlay\", \"tracks\": [{\"data\": {\"type\": \"multivec\", \"url\": \"https://server.gosling-lang.org/api/v1/tileset_info/?d=cistrome-multivec\", \"categories\": [\"sample 1\", \"sample 2\", \"sample 3\", \"sample 4\"], \"column\": \"position\", \"row\": \"sample\", \"value\": \"peak\"}, \"mark\": \"bar\", \"color\": {\"field\": \"sample\", \"type\": \"nominal\"}, \"layout\": \"linear\", \"row\": {\"field\": \"sample\", \"type\": \"nominal\"}, \"static\": true, \"x\": {\"domain\": {\"chromosome\": \"chr1\"}, \"field\": \"start\", \"type\": \"genomic\"}, \"xe\": {\"field\": \"end\", \"type\": \"genomic\"}, \"y\": {\"field\": \"peak\", \"type\": \"quantitative\"}}, {\"data\": {\"type\": \"multivec\", \"url\": \"https://server.gosling-lang.org/api/v1/tileset_info/?d=cistrome-multivec\", \"categories\": [\"sample 1\", \"sample 2\", \"sample 3\", \"sample 4\"], \"column\": \"position\", \"row\": \"sample\", \"value\": \"peak\"}, \"mark\": \"brush\", \"color\": {\"value\": \"steelblue\"}, \"row\": {\"field\": \"sample\", \"type\": \"nominal\"}, \"x\": {\"field\": \"start\", \"linkingId\": \"brushId\", \"type\": \"genomic\"}, \"xe\": {\"field\": \"end\", \"type\": \"genomic\"}, \"y\": {\"field\": \"peak\", \"type\": \"quantitative\"}}], \"height\": 200, \"width\": 400}], \"spacing\": 40}, {\"tracks\": [{\"data\": {\"type\": \"multivec\", \"url\": \"https://server.gosling-lang.org/api/v1/tileset_info/?d=cistrome-multivec\", \"categories\": [\"sample 1\", \"sample 2\", \"sample 3\", \"sample 4\"], \"column\": \"position\", \"row\": \"sample\", \"value\": \"peak\"}, \"mark\": \"bar\", \"color\": {\"field\": \"sample\", \"type\": \"nominal\"}, \"height\": 200, \"row\": {\"field\": \"sample\", \"type\": \"nominal\"}, \"style\": {\"background\": \"steelblue\", \"backgroundOpacity\": 0.2}, \"title\": \"Detailed View\", \"width\": 690, \"x\": {\"domain\": {\"chromosome\": \"chr1\", \"interval\": [160000000, 200000000]}, \"field\": \"start\", \"linkingId\": \"brushId\", \"type\": \"genomic\"}, \"xe\": {\"field\": \"end\", \"type\": \"genomic\"}, \"y\": {\"field\": \"peak\", \"type\": \"quantitative\"}}]}], \"title\": \"Recreation of Example Visual\", \"subtitle\": \"Drag the brush to get a detailed view\"};\n",
       "    let opt = {\"padding\": 0, \"theme\": null};\n",
       "    gosling.embed(el, spec, opt).catch((err) => {\n",
       "      el.innerHTML = `    <div class=\"error\">\n",
       "        <p>JavaScript Error: ${error.message}</p>\n",
       "        <p>This usually means there's a typo in your Gosling specification. See the javascript console for the full traceback.</p>\n",
       "    </div>`;\n",
       "      throw error;\n",
       "    });\n",
       "  </script>\n",
       "</body>\n",
       "</html>"
      ],
      "text/plain": [
       "View({\n",
       "  arrangement: 'vertical',\n",
       "  subtitle: 'Drag the brush to get a detailed view',\n",
       "  title: 'Recreation of Example Visual',\n",
       "  views: (View({\n",
       "    arrangement: 'horizontal',\n",
       "    spacing: 40,\n",
       "    views: (View({\n",
       "      alignment: 'overlay',\n",
       "      centerRadius: 0.4,\n",
       "      height: 130,\n",
       "      layout: 'circular',\n",
       "      tracks: (Track({\n",
       "        color: Color({\n",
       "          field: 'sample',\n",
       "          type: 'nominal'\n",
       "        }),\n",
       "        data: MultivecData({\n",
       "          categories: ['sample 1', 'sample 2', 'sample 3', 'sample 4'],\n",
       "          column: 'position',\n",
       "          row: 'sample',\n",
       "          type: 'multivec',\n",
       "          url: 'https://server.gosling-lang.org/api/v1/tileset_info/?d=cistrome-multivec',\n",
       "          value: 'peak'\n",
       "        }),\n",
       "        layout: 'circular',\n",
       "        mark: 'bar',\n",
       "        row: Row({\n",
       "          field: 'sample',\n",
       "          type: 'nominal'\n",
       "        }),\n",
       "        static: True,\n",
       "        x: X({\n",
       "          domain: GenomicDomain(DomainChr({\n",
       "            chromosome: 'chr1'\n",
       "          })),\n",
       "          field: 'start',\n",
       "          type: 'genomic'\n",
       "        }),\n",
       "        xe: Xe({\n",
       "          field: 'end',\n",
       "          type: 'genomic'\n",
       "        }),\n",
       "        y: Y({\n",
       "          field: 'peak',\n",
       "          type: 'quantitative'\n",
       "        })\n",
       "      }), Track({\n",
       "        color: ColorValue({\n",
       "          value: 'steelblue'\n",
       "        }),\n",
       "        data: MultivecData({\n",
       "          categories: ['sample 1', 'sample 2', 'sample 3', 'sample 4'],\n",
       "          column: 'position',\n",
       "          row: 'sample',\n",
       "          type: 'multivec',\n",
       "          url: 'https://server.gosling-lang.org/api/v1/tileset_info/?d=cistrome-multivec',\n",
       "          value: 'peak'\n",
       "        }),\n",
       "        mark: 'brush',\n",
       "        row: Row({\n",
       "          field: 'sample',\n",
       "          type: 'nominal'\n",
       "        }),\n",
       "        x: X({\n",
       "          field: 'start',\n",
       "          linkingId: 'brushId',\n",
       "          type: 'genomic'\n",
       "        }),\n",
       "        xe: Xe({\n",
       "          field: 'end',\n",
       "          type: 'genomic'\n",
       "        }),\n",
       "        y: Y({\n",
       "          field: 'peak',\n",
       "          type: 'quantitative'\n",
       "        })\n",
       "      })),\n",
       "      width: 250\n",
       "    }), View({\n",
       "      alignment: 'overlay',\n",
       "      height: 200,\n",
       "      tracks: (Track({\n",
       "        color: Color({\n",
       "          field: 'sample',\n",
       "          type: 'nominal'\n",
       "        }),\n",
       "        data: MultivecData({\n",
       "          categories: ['sample 1', 'sample 2', 'sample 3', 'sample 4'],\n",
       "          column: 'position',\n",
       "          row: 'sample',\n",
       "          type: 'multivec',\n",
       "          url: 'https://server.gosling-lang.org/api/v1/tileset_info/?d=cistrome-multivec',\n",
       "          value: 'peak'\n",
       "        }),\n",
       "        layout: 'linear',\n",
       "        mark: 'bar',\n",
       "        row: Row({\n",
       "          field: 'sample',\n",
       "          type: 'nominal'\n",
       "        }),\n",
       "        static: True,\n",
       "        x: X({\n",
       "          domain: GenomicDomain(DomainChr({\n",
       "            chromosome: 'chr1'\n",
       "          })),\n",
       "          field: 'start',\n",
       "          type: 'genomic'\n",
       "        }),\n",
       "        xe: Xe({\n",
       "          field: 'end',\n",
       "          type: 'genomic'\n",
       "        }),\n",
       "        y: Y({\n",
       "          field: 'peak',\n",
       "          type: 'quantitative'\n",
       "        })\n",
       "      }), Track({\n",
       "        color: ColorValue({\n",
       "          value: 'steelblue'\n",
       "        }),\n",
       "        data: MultivecData({\n",
       "          categories: ['sample 1', 'sample 2', 'sample 3', 'sample 4'],\n",
       "          column: 'position',\n",
       "          row: 'sample',\n",
       "          type: 'multivec',\n",
       "          url: 'https://server.gosling-lang.org/api/v1/tileset_info/?d=cistrome-multivec',\n",
       "          value: 'peak'\n",
       "        }),\n",
       "        mark: 'brush',\n",
       "        row: Row({\n",
       "          field: 'sample',\n",
       "          type: 'nominal'\n",
       "        }),\n",
       "        x: X({\n",
       "          field: 'start',\n",
       "          linkingId: 'brushId',\n",
       "          type: 'genomic'\n",
       "        }),\n",
       "        xe: Xe({\n",
       "          field: 'end',\n",
       "          type: 'genomic'\n",
       "        }),\n",
       "        y: Y({\n",
       "          field: 'peak',\n",
       "          type: 'quantitative'\n",
       "        })\n",
       "      })),\n",
       "      width: 400\n",
       "    }))\n",
       "  }), View({\n",
       "    tracks: [Track({\n",
       "      color: Color({\n",
       "        field: 'sample',\n",
       "        type: 'nominal'\n",
       "      }),\n",
       "      data: MultivecData({\n",
       "        categories: ['sample 1', 'sample 2', 'sample 3', 'sample 4'],\n",
       "        column: 'position',\n",
       "        row: 'sample',\n",
       "        type: 'multivec',\n",
       "        url: 'https://server.gosling-lang.org/api/v1/tileset_info/?d=cistrome-multivec',\n",
       "        value: 'peak'\n",
       "      }),\n",
       "      height: 200,\n",
       "      mark: 'bar',\n",
       "      row: Row({\n",
       "        field: 'sample',\n",
       "        type: 'nominal'\n",
       "      }),\n",
       "      style: Style({\n",
       "        background: 'steelblue',\n",
       "        backgroundOpacity: 0.2\n",
       "      }),\n",
       "      title: 'Detailed View',\n",
       "      width: 690,\n",
       "      x: X({\n",
       "        domain: GenomicDomain(DomainChrInterval({\n",
       "          chromosome: 'chr1',\n",
       "          interval: [160000000, 200000000]\n",
       "        })),\n",
       "        field: 'start',\n",
       "        linkingId: 'brushId',\n",
       "        type: 'genomic'\n",
       "      }),\n",
       "      xe: Xe({\n",
       "        field: 'end',\n",
       "        type: 'genomic'\n",
       "      }),\n",
       "      y: Y({\n",
       "        field: 'peak',\n",
       "        type: 'quantitative'\n",
       "      })\n",
       "    })]\n",
       "  }))\n",
       "})"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "aligned_views=gos.horizontal(\n",
    "    circ_overview,\n",
    "    lin_overview,\n",
    "    spacing=40\n",
    ")\n",
    "final_view=gos.vertical(\n",
    "    aligned_views,\n",
    "    detailed_track\n",
    ").properties(\n",
    "    title=\"Recreation of Example Visual\",\n",
    "    subtitle=\"Drag the brush to get a detailed view\"\n",
    ")\n",
    "\n",
    "final_view.save('visuals/recreation/recreated_view.json')\n",
    "display(final_view)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4. Comparative Heatmaps\n",
    "In this section, I am creating [this](https://gosling.js.org/?example=MATRIX_HFFC6) visual."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "# define each data point\n",
    "\n",
    "# genes included in view\n",
    "genes = gos.beddb(\n",
    "    url=\"https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation\",\n",
    "    \n",
    "    # create genomic fields\n",
    "    genomicFields=[\n",
    "        {\"index\": 1, \"name\": \"start\"},\n",
    "        {\"index\": 2, \"name\": \"end\"},\n",
    "    ],\n",
    "    # create values for each\n",
    "    valueFields=[\n",
    "        {\"index\": 5, \"name\": \"strand\", \"type\": \"nominal\"},\n",
    "        {\"index\": 3, \"name\": \"name\", \"type\": \"nominal\"},\n",
    "    ],\n",
    ")\n",
    "# 3 views on top of each heatmap\n",
    "h3k4 = gos.bigwig(\n",
    "    url=\"https://s3.amazonaws.com/gosling-lang.org/data/HFFc6_H3K4me3.bigWig\",\n",
    "    column=\"position\",\n",
    "    value=\"peak\",\n",
    ")\n",
    "\n",
    "atac = gos.bigwig(\n",
    "    url=\"https://s3.amazonaws.com/gosling-lang.org/data/HFFc6_Atacseq.mRp.clN.bigWig\",\n",
    "    column=\"position\",\n",
    "    value=\"peak\",\n",
    ")\n",
    "\n",
    "ctcf = gos.bigwig(\n",
    "    url=\"https://s3.amazonaws.com/gosling-lang.org/data/HFFC6_CTCF.mRp.clN.bigWig\",\n",
    "    column=\"position\",\n",
    "    value=\"peak\",\n",
    ")\n",
    "\n",
    "# micro c and hi c data --> matrix type\n",
    "micro_c = gos.matrix(\"https://server.gosling-lang.org/api/v1/tileset_info/?d=hffc6-microc-hg38\")\n",
    "hi_c = gos.matrix(\"https://server.gosling-lang.org/api/v1/tileset_info/?d=hffc6-hic-hg38\")\n",
    "\n",
    "# bottom multivec data\n",
    "epilogos = gos.multivec(\n",
    "    url=\"https://server.gosling-lang.org/api/v1/tileset_info/?d=epilogos-hg38\",\n",
    "    row=\"category\",\n",
    "    column=\"position\",\n",
    "    value=\"value\",\n",
    "    categories=[\n",
    "        \"Active TSS\", \"Flanking Active TSS\", \"Transcr at gene 5\\\\' and 3\\\\'\",\n",
    "        \"Strong transcription\", \"Weak transcription\", \"Genic enhancers\",\n",
    "        \"Enhancers\", \"ZNF genes & repeats\", \"Heterochromatin\",\n",
    "        \"Bivalent/Poised TSS\", \"Flanking Bivalent TSS/Enh\", \"Bivalent Enhancer\",\n",
    "        \"Repressed PolyComb\", \"Weak Repressed PolyComb\", \"Quiescent/Low\",\n",
    "    ],\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we make the three top tracks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "h3k4_track = gos.Track(\n",
    "    data=h3k4,\n",
    ").mark_bar().encode(\n",
    "    x=gos.X('start:G', axis='top'),\n",
    "    xe='end:G',\n",
    "    y=gos.Y(\"peak:Q\"),\n",
    "    color=gos.value(\"darkgreen\"),\n",
    ").properties(title=\"HFFc6_H3K4me3\", height=375, width=375/15)\n",
    "\n",
    "atac_track = gos.Track(atac).mark_bar().encode(\n",
    "    x=\"start:G\",\n",
    "    xe=\"end:G\",\n",
    "    y=gos.Y(\"peak:Q\", axis=\"none\"),\n",
    "    color=gos.value(\"#E79F00\"),\n",
    ").properties(title=\"HFFc6_ATAC\", height=375, width=375/15)\n",
    "\n",
    "# create gene annotation base with ctcf\n",
    "gene_base = gos.Track(genes).encode(\n",
    "    x=\"start:G\",\n",
    "    size=gos.value(13),\n",
    "    stroke=gos.value(\"white\"),\n",
    "    strokeWidth=gos.value(1),\n",
    "    color=gos.value(\"#CB7AA7\"),\n",
    "    # sets \n",
    "    row=gos.Row(\"strand:N\", domain=[\"+\", \"-\"]),\n",
    ")\n",
    "\n",
    "# create overlay atop gene base\n",
    "gene_overlay = gos.overlay(\n",
    "    gos.Track(ctcf).mark_bar().encode(\n",
    "        x=\"start:G\",\n",
    "        xe=\"end:G\",\n",
    "        y=gos.Y(\"peak:Q\",axis='none'),\n",
    "        color=gos.value(\"#0072B2\")\n",
    "    ),\n",
    "    # add a triangle right for any positive values\n",
    "    gene_base.mark_triangleRight(backgroundOpacity=0).encode(\n",
    "        color=gos.value(\"#CB7AA7\"),\n",
    "    ).transform_filter(\"strand\", oneOf=[\"+\"]),\n",
    "    # add left triangle for negative values\n",
    "    gene_base.mark_triangleLeft(backgroundOpacity=0).encode(\n",
    "        color=gos.value(\"#029F73\"),\n",
    "    ).transform_filter(\"strand\", oneOf=[\"-\"]).properties(title=\"HFFC6_CTCF\")\n",
    "\n",
    ").properties(height=375, width=375/15)\n",
    "\n",
    "\n",
    "# stack 3 top views\n",
    "left_stack = gos.stack(h3k4_track, atac_track, gene_overlay).properties(\n",
    "    orientation=\"vertical\",\n",
    "    yOffset=375 / 2.78,\n",
    ")\n",
    "# change top_stack dims (otherwise very long and skinny)\n",
    "top_stack = gos.stack(\n",
    "    h3k4_track.properties(height=375/15, width=375), \n",
    "    atac_track.properties(height=375/15, width=375), \n",
    "    gene_overlay.properties(height=375/15, width=375)\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Add the hic and micro c matrices"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<!DOCTYPE html>\n",
       "<html>\n",
       "<head>\n",
       "  <style>.error { color: red; }</style>\n",
       "</head>\n",
       "<body>\n",
       "  <div id=\"jupyter-gosling-f02c88cc32f74ffab9587c18e8c7d0b4\"></div>\n",
       "  <script type=\"module\">\n",
       "    import { importWithMap } from \"https://unpkg.com/dynamic-importmap@0.1.0\";\n",
       "    let importMap = {\"imports\": {\"react\": \"https://esm.sh/react@18\", \"react-dom\": \"https://esm.sh/react-dom@18\", \"pixi\": \"https://esm.sh/pixi.js@6?bundle\", \"higlass\": \"https://esm.sh/higlass@1.13?external=react,react-dom,pixi&bundle\", \"gosling.js\": \"https://esm.sh/gosling.js@1.0.2?external=react,react-dom,pixi,higlass&bundle\"}};\n",
       "    let gosling = await importWithMap(\"gosling.js\", importMap);\n",
       "    let el = document.getElementById('jupyter-gosling-f02c88cc32f74ffab9587c18e8c7d0b4');\n",
       "    let spec = {\"arrangement\": \"horizontal\", \"views\": [{\"alignment\": \"stack\", \"tracks\": [{\"data\": {\"type\": \"bigwig\", \"url\": \"https://s3.amazonaws.com/gosling-lang.org/data/HFFc6_H3K4me3.bigWig\", \"column\": \"position\", \"value\": \"peak\"}, \"mark\": \"bar\", \"color\": {\"value\": \"darkgreen\"}, \"height\": 375, \"title\": \"HFFc6_H3K4me3\", \"width\": 25.0, \"x\": {\"axis\": \"top\", \"field\": \"start\", \"type\": \"genomic\"}, \"xe\": {\"field\": \"end\", \"type\": \"genomic\"}, \"y\": {\"field\": \"peak\", \"type\": \"quantitative\"}}, {\"data\": {\"type\": \"bigwig\", \"url\": \"https://s3.amazonaws.com/gosling-lang.org/data/HFFc6_Atacseq.mRp.clN.bigWig\", \"column\": \"position\", \"value\": \"peak\"}, \"mark\": \"bar\", \"color\": {\"value\": \"#E79F00\"}, \"height\": 375, \"title\": \"HFFc6_ATAC\", \"width\": 25.0, \"x\": {\"field\": \"start\", \"type\": \"genomic\"}, \"xe\": {\"field\": \"end\", \"type\": \"genomic\"}, \"y\": {\"axis\": \"none\", \"field\": \"peak\", \"type\": \"quantitative\"}}, {\"alignment\": \"overlay\", \"tracks\": [{\"data\": {\"type\": \"bigwig\", \"url\": \"https://s3.amazonaws.com/gosling-lang.org/data/HFFC6_CTCF.mRp.clN.bigWig\", \"column\": \"position\", \"value\": \"peak\"}, \"mark\": \"bar\", \"color\": {\"value\": \"#0072B2\"}, \"x\": {\"field\": \"start\", \"type\": \"genomic\"}, \"xe\": {\"field\": \"end\", \"type\": \"genomic\"}, \"y\": {\"axis\": \"none\", \"field\": \"peak\", \"type\": \"quantitative\"}}, {\"data\": {\"type\": \"beddb\", \"url\": \"https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation\", \"genomicFields\": [{\"index\": 1, \"name\": \"start\"}, {\"index\": 2, \"name\": \"end\"}], \"valueFields\": [{\"index\": 5, \"name\": \"strand\", \"type\": \"nominal\"}, {\"index\": 3, \"name\": \"name\", \"type\": \"nominal\"}]}, \"mark\": \"triangleRight\", \"color\": {\"value\": \"#CB7AA7\"}, \"dataTransform\": [{\"type\": \"filter\", \"field\": \"strand\", \"oneOf\": [\"+\"]}], \"row\": {\"domain\": [\"+\", \"-\"], \"field\": \"strand\", \"type\": \"nominal\"}, \"size\": {\"value\": 13}, \"stroke\": {\"value\": \"white\"}, \"strokeWidth\": {\"value\": 1}, \"style\": {\"backgroundOpacity\": 0}, \"x\": {\"field\": \"start\", \"type\": \"genomic\"}}, {\"data\": {\"type\": \"beddb\", \"url\": \"https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation\", \"genomicFields\": [{\"index\": 1, \"name\": \"start\"}, {\"index\": 2, \"name\": \"end\"}], \"valueFields\": [{\"index\": 5, \"name\": \"strand\", \"type\": \"nominal\"}, {\"index\": 3, \"name\": \"name\", \"type\": \"nominal\"}]}, \"mark\": \"triangleLeft\", \"color\": {\"value\": \"#029F73\"}, \"dataTransform\": [{\"type\": \"filter\", \"field\": \"strand\", \"oneOf\": [\"-\"]}], \"row\": {\"domain\": [\"+\", \"-\"], \"field\": \"strand\", \"type\": \"nominal\"}, \"size\": {\"value\": 13}, \"stroke\": {\"value\": \"white\"}, \"strokeWidth\": {\"value\": 1}, \"style\": {\"backgroundOpacity\": 0}, \"title\": \"HFFC6_CTCF\", \"x\": {\"field\": \"start\", \"type\": \"genomic\"}}], \"height\": 375, \"width\": 25.0}], \"orientation\": \"vertical\", \"yOffset\": 134.89208633093526}, {\"views\": [{\"arrangement\": \"vertical\", \"views\": [{\"alignment\": \"stack\", \"tracks\": [{\"data\": {\"type\": \"bigwig\", \"url\": \"https://s3.amazonaws.com/gosling-lang.org/data/HFFc6_H3K4me3.bigWig\", \"column\": \"position\", \"value\": \"peak\"}, \"mark\": \"bar\", \"color\": {\"value\": \"darkgreen\"}, \"height\": 25.0, \"title\": \"HFFc6_H3K4me3\", \"width\": 375, \"x\": {\"axis\": \"top\", \"field\": \"start\", \"type\": \"genomic\"}, \"xe\": {\"field\": \"end\", \"type\": \"genomic\"}, \"y\": {\"field\": \"peak\", \"type\": \"quantitative\"}}, {\"data\": {\"type\": \"bigwig\", \"url\": \"https://s3.amazonaws.com/gosling-lang.org/data/HFFc6_Atacseq.mRp.clN.bigWig\", \"column\": \"position\", \"value\": \"peak\"}, \"mark\": \"bar\", \"color\": {\"value\": \"#E79F00\"}, \"height\": 25.0, \"title\": \"HFFc6_ATAC\", \"width\": 375, \"x\": {\"field\": \"start\", \"type\": \"genomic\"}, \"xe\": {\"field\": \"end\", \"type\": \"genomic\"}, \"y\": {\"axis\": \"none\", \"field\": \"peak\", \"type\": \"quantitative\"}}, {\"alignment\": \"overlay\", \"tracks\": [{\"data\": {\"type\": \"bigwig\", \"url\": \"https://s3.amazonaws.com/gosling-lang.org/data/HFFC6_CTCF.mRp.clN.bigWig\", \"column\": \"position\", \"value\": \"peak\"}, \"mark\": \"bar\", \"color\": {\"value\": \"#0072B2\"}, \"x\": {\"field\": \"start\", \"type\": \"genomic\"}, \"xe\": {\"field\": \"end\", \"type\": \"genomic\"}, \"y\": {\"axis\": \"none\", \"field\": \"peak\", \"type\": \"quantitative\"}}, {\"data\": {\"type\": \"beddb\", \"url\": \"https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation\", \"genomicFields\": [{\"index\": 1, \"name\": \"start\"}, {\"index\": 2, \"name\": \"end\"}], \"valueFields\": [{\"index\": 5, \"name\": \"strand\", \"type\": \"nominal\"}, {\"index\": 3, \"name\": \"name\", \"type\": \"nominal\"}]}, \"mark\": \"triangleRight\", \"color\": {\"value\": \"#CB7AA7\"}, \"dataTransform\": [{\"type\": \"filter\", \"field\": \"strand\", \"oneOf\": [\"+\"]}], \"row\": {\"domain\": [\"+\", \"-\"], \"field\": \"strand\", \"type\": \"nominal\"}, \"size\": {\"value\": 13}, \"stroke\": {\"value\": \"white\"}, \"strokeWidth\": {\"value\": 1}, \"style\": {\"backgroundOpacity\": 0}, \"x\": {\"field\": \"start\", \"type\": \"genomic\"}}, {\"data\": {\"type\": \"beddb\", \"url\": \"https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation\", \"genomicFields\": [{\"index\": 1, \"name\": \"start\"}, {\"index\": 2, \"name\": \"end\"}], \"valueFields\": [{\"index\": 5, \"name\": \"strand\", \"type\": \"nominal\"}, {\"index\": 3, \"name\": \"name\", \"type\": \"nominal\"}]}, \"mark\": \"triangleLeft\", \"color\": {\"value\": \"#029F73\"}, \"dataTransform\": [{\"type\": \"filter\", \"field\": \"strand\", \"oneOf\": [\"-\"]}], \"row\": {\"domain\": [\"+\", \"-\"], \"field\": \"strand\", \"type\": \"nominal\"}, \"size\": {\"value\": 13}, \"stroke\": {\"value\": \"white\"}, \"strokeWidth\": {\"value\": 1}, \"style\": {\"backgroundOpacity\": 0}, \"title\": \"HFFC6_CTCF\", \"x\": {\"field\": \"start\", \"type\": \"genomic\"}}], \"height\": 25.0, \"width\": 375}]}, {\"tracks\": [{\"data\": {\"type\": \"matrix\", \"url\": \"https://server.gosling-lang.org/api/v1/tileset_info/?d=hffc6-microc-hg38\"}, \"mark\": \"rect\", \"color\": {\"field\": \"value\", \"range\": \"warm\", \"type\": \"quantitative\"}, \"height\": 375, \"title\": \"HFFc6_Micro-C\", \"width\": 375, \"x\": {\"field\": \"xs\", \"type\": \"genomic\"}, \"xe\": {\"field\": \"xe\", \"type\": \"genomic\"}, \"y\": {\"field\": \"ys\", \"type\": \"genomic\"}, \"ye\": {\"field\": \"ye\", \"type\": \"genomic\"}}]}, {\"tracks\": [{\"data\": {\"type\": \"multivec\", \"url\": \"https://server.gosling-lang.org/api/v1/tileset_info/?d=epilogos-hg38\", \"row\": \"category\", \"column\": \"position\", \"value\": \"value\", \"categories\": [\"Active TSS\", \"Flanking Active TSS\", \"Transcr at gene 5\\\\' and 3\\\\'\", \"Strong transcription\", \"Weak transcription\", \"Genic enhancers\", \"Enhancers\", \"ZNF genes & repeats\", \"Heterochromatin\", \"Bivalent/Poised TSS\", \"Flanking Bivalent TSS/Enh\", \"Bivalent Enhancer\", \"Repressed PolyComb\", \"Weak Repressed PolyComb\", \"Quiescent/Low\"]}, \"mark\": \"bar\", \"color\": {\"field\": \"category\", \"range\": [\"#FF0000\", \"#FF4500\", \"#32CD32\", \"#008000\", \"#006400\", \"#C2E105\", \"#FFFF00\", \"#66CDAA\", \"#8A91D0\", \"#CD5C5C\", \"#E9967A\", \"#BDB76B\", \"#808080\", \"#C0C0C0\", \"gray\"], \"type\": \"nominal\"}, \"dataTransform\": [{\"type\": \"filter\", \"field\": \"value\", \"inRange\": [0, 999999]}], \"height\": 25.0, \"title\": \"Epilogos (hg38)\", \"width\": 375, \"x\": {\"axis\": \"none\", \"field\": \"start\", \"type\": \"genomic\"}, \"xe\": {\"field\": \"end\", \"type\": \"genomic\"}, \"y\": {\"axis\": \"none\", \"field\": \"value\", \"type\": \"quantitative\"}}]}], \"spacing\": 0}, {\"arrangement\": \"vertical\", \"views\": [{\"alignment\": \"stack\", \"tracks\": [{\"data\": {\"type\": \"bigwig\", \"url\": \"https://s3.amazonaws.com/gosling-lang.org/data/HFFc6_H3K4me3.bigWig\", \"column\": \"position\", \"value\": \"peak\"}, \"mark\": \"bar\", \"color\": {\"value\": \"darkgreen\"}, \"height\": 25.0, \"title\": \"HFFc6_H3K4me3\", \"width\": 375, \"x\": {\"axis\": \"top\", \"field\": \"start\", \"type\": \"genomic\"}, \"xe\": {\"field\": \"end\", \"type\": \"genomic\"}, \"y\": {\"field\": \"peak\", \"type\": \"quantitative\"}}, {\"data\": {\"type\": \"bigwig\", \"url\": \"https://s3.amazonaws.com/gosling-lang.org/data/HFFc6_Atacseq.mRp.clN.bigWig\", \"column\": \"position\", \"value\": \"peak\"}, \"mark\": \"bar\", \"color\": {\"value\": \"#E79F00\"}, \"height\": 25.0, \"title\": \"HFFc6_ATAC\", \"width\": 375, \"x\": {\"field\": \"start\", \"type\": \"genomic\"}, \"xe\": {\"field\": \"end\", \"type\": \"genomic\"}, \"y\": {\"axis\": \"none\", \"field\": \"peak\", \"type\": \"quantitative\"}}, {\"alignment\": \"overlay\", \"tracks\": [{\"data\": {\"type\": \"bigwig\", \"url\": \"https://s3.amazonaws.com/gosling-lang.org/data/HFFC6_CTCF.mRp.clN.bigWig\", \"column\": \"position\", \"value\": \"peak\"}, \"mark\": \"bar\", \"color\": {\"value\": \"#0072B2\"}, \"x\": {\"field\": \"start\", \"type\": \"genomic\"}, \"xe\": {\"field\": \"end\", \"type\": \"genomic\"}, \"y\": {\"axis\": \"none\", \"field\": \"peak\", \"type\": \"quantitative\"}}, {\"data\": {\"type\": \"beddb\", \"url\": \"https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation\", \"genomicFields\": [{\"index\": 1, \"name\": \"start\"}, {\"index\": 2, \"name\": \"end\"}], \"valueFields\": [{\"index\": 5, \"name\": \"strand\", \"type\": \"nominal\"}, {\"index\": 3, \"name\": \"name\", \"type\": \"nominal\"}]}, \"mark\": \"triangleRight\", \"color\": {\"value\": \"#CB7AA7\"}, \"dataTransform\": [{\"type\": \"filter\", \"field\": \"strand\", \"oneOf\": [\"+\"]}], \"row\": {\"domain\": [\"+\", \"-\"], \"field\": \"strand\", \"type\": \"nominal\"}, \"size\": {\"value\": 13}, \"stroke\": {\"value\": \"white\"}, \"strokeWidth\": {\"value\": 1}, \"style\": {\"backgroundOpacity\": 0}, \"x\": {\"field\": \"start\", \"type\": \"genomic\"}}, {\"data\": {\"type\": \"beddb\", \"url\": \"https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation\", \"genomicFields\": [{\"index\": 1, \"name\": \"start\"}, {\"index\": 2, \"name\": \"end\"}], \"valueFields\": [{\"index\": 5, \"name\": \"strand\", \"type\": \"nominal\"}, {\"index\": 3, \"name\": \"name\", \"type\": \"nominal\"}]}, \"mark\": \"triangleLeft\", \"color\": {\"value\": \"#029F73\"}, \"dataTransform\": [{\"type\": \"filter\", \"field\": \"strand\", \"oneOf\": [\"-\"]}], \"row\": {\"domain\": [\"+\", \"-\"], \"field\": \"strand\", \"type\": \"nominal\"}, \"size\": {\"value\": 13}, \"stroke\": {\"value\": \"white\"}, \"strokeWidth\": {\"value\": 1}, \"style\": {\"backgroundOpacity\": 0}, \"title\": \"HFFC6_CTCF\", \"x\": {\"field\": \"start\", \"type\": \"genomic\"}}], \"height\": 25.0, \"width\": 375}]}, {\"tracks\": [{\"data\": {\"type\": \"matrix\", \"url\": \"https://server.gosling-lang.org/api/v1/tileset_info/?d=hffc6-hic-hg38\"}, \"mark\": \"rect\", \"color\": {\"field\": \"value\", \"range\": \"warm\", \"type\": \"quantitative\"}, \"height\": 375, \"title\": \"HFFc6_Hi-C\", \"width\": 375, \"x\": {\"field\": \"xs\", \"type\": \"genomic\"}, \"xe\": {\"field\": \"xe\", \"type\": \"genomic\"}, \"y\": {\"field\": \"ys\", \"type\": \"genomic\"}, \"ye\": {\"field\": \"ye\", \"type\": \"genomic\"}}]}, {\"tracks\": [{\"data\": {\"type\": \"multivec\", \"url\": \"https://server.gosling-lang.org/api/v1/tileset_info/?d=epilogos-hg38\", \"row\": \"category\", \"column\": \"position\", \"value\": \"value\", \"categories\": [\"Active TSS\", \"Flanking Active TSS\", \"Transcr at gene 5\\\\' and 3\\\\'\", \"Strong transcription\", \"Weak transcription\", \"Genic enhancers\", \"Enhancers\", \"ZNF genes & repeats\", \"Heterochromatin\", \"Bivalent/Poised TSS\", \"Flanking Bivalent TSS/Enh\", \"Bivalent Enhancer\", \"Repressed PolyComb\", \"Weak Repressed PolyComb\", \"Quiescent/Low\"]}, \"mark\": \"bar\", \"color\": {\"field\": \"category\", \"range\": [\"#FF0000\", \"#FF4500\", \"#32CD32\", \"#008000\", \"#006400\", \"#C2E105\", \"#FFFF00\", \"#66CDAA\", \"#8A91D0\", \"#CD5C5C\", \"#E9967A\", \"#BDB76B\", \"#808080\", \"#C0C0C0\", \"gray\"], \"type\": \"nominal\"}, \"dataTransform\": [{\"type\": \"filter\", \"field\": \"value\", \"inRange\": [0, 999999]}], \"height\": 25.0, \"title\": \"Epilogos (hg38)\", \"width\": 375, \"x\": {\"axis\": \"none\", \"field\": \"start\", \"type\": \"genomic\"}, \"xe\": {\"field\": \"end\", \"type\": \"genomic\"}, \"y\": {\"axis\": \"none\", \"field\": \"value\", \"type\": \"quantitative\"}}]}], \"spacing\": 0}], \"spacing\": 30}], \"title\": \"Matrix Visualization Recreation\", \"xDomain\": {\"chromosome\": \"chr7\", \"interval\": [77700000, 81000000]}, \"spacing\": 1, \"linkingId\": \"-\"};\n",
       "    let opt = {\"padding\": 0, \"theme\": null};\n",
       "    gosling.embed(el, spec, opt).catch((err) => {\n",
       "      el.innerHTML = `    <div class=\"error\">\n",
       "        <p>JavaScript Error: ${error.message}</p>\n",
       "        <p>This usually means there's a typo in your Gosling specification. See the javascript console for the full traceback.</p>\n",
       "    </div>`;\n",
       "      throw error;\n",
       "    });\n",
       "  </script>\n",
       "</body>\n",
       "</html>"
      ],
      "text/plain": [
       "View({\n",
       "  arrangement: 'horizontal',\n",
       "  linkingId: '-',\n",
       "  spacing: 1,\n",
       "  title: 'Matrix Visualization Recreation',\n",
       "  views: (View({\n",
       "    alignment: 'stack',\n",
       "    orientation: 'vertical',\n",
       "    tracks: (Track({\n",
       "      color: ColorValue({\n",
       "        value: 'darkgreen'\n",
       "      }),\n",
       "      data: {'type': 'bigwig', 'url': 'https://s3.amazonaws.com/gosling-lang.org/data/HFFc6_H3K4me3.bigWig', 'column': 'position', 'value': 'peak'},\n",
       "      height: 375,\n",
       "      mark: 'bar',\n",
       "      title: 'HFFc6_H3K4me3',\n",
       "      width: 25.0,\n",
       "      x: X({\n",
       "        axis: 'top',\n",
       "        field: 'start',\n",
       "        type: 'genomic'\n",
       "      }),\n",
       "      xe: Xe({\n",
       "        field: 'end',\n",
       "        type: 'genomic'\n",
       "      }),\n",
       "      y: Y({\n",
       "        field: 'peak',\n",
       "        type: 'quantitative'\n",
       "      })\n",
       "    }), Track({\n",
       "      color: ColorValue({\n",
       "        value: '#E79F00'\n",
       "      }),\n",
       "      data: {'type': 'bigwig', 'url': 'https://s3.amazonaws.com/gosling-lang.org/data/HFFc6_Atacseq.mRp.clN.bigWig', 'column': 'position', 'value': 'peak'},\n",
       "      height: 375,\n",
       "      mark: 'bar',\n",
       "      title: 'HFFc6_ATAC',\n",
       "      width: 25.0,\n",
       "      x: X({\n",
       "        field: 'start',\n",
       "        type: 'genomic'\n",
       "      }),\n",
       "      xe: Xe({\n",
       "        field: 'end',\n",
       "        type: 'genomic'\n",
       "      }),\n",
       "      y: Y({\n",
       "        axis: 'none',\n",
       "        field: 'peak',\n",
       "        type: 'quantitative'\n",
       "      })\n",
       "    }), View({\n",
       "      alignment: 'overlay',\n",
       "      height: 375,\n",
       "      tracks: (Track({\n",
       "        color: ColorValue({\n",
       "          value: '#0072B2'\n",
       "        }),\n",
       "        data: {'type': 'bigwig', 'url': 'https://s3.amazonaws.com/gosling-lang.org/data/HFFC6_CTCF.mRp.clN.bigWig', 'column': 'position', 'value': 'peak'},\n",
       "        mark: 'bar',\n",
       "        x: X({\n",
       "          field: 'start',\n",
       "          type: 'genomic'\n",
       "        }),\n",
       "        xe: Xe({\n",
       "          field: 'end',\n",
       "          type: 'genomic'\n",
       "        }),\n",
       "        y: Y({\n",
       "          axis: 'none',\n",
       "          field: 'peak',\n",
       "          type: 'quantitative'\n",
       "        })\n",
       "      }), Track({\n",
       "        color: ColorValue({\n",
       "          value: '#CB7AA7'\n",
       "        }),\n",
       "        data: {'type': 'beddb', 'url': 'https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation', 'genomicFields': [{'index': 1, 'name': 'start'}, {'index': 2, 'name': 'end'}], 'valueFields': [{'index': 5, 'name': 'strand', 'type': 'nominal'}, {'index': 3, 'name': 'name', 'type': 'nominal'}]},\n",
       "        dataTransform: [FilterTransform({\n",
       "          field: 'strand',\n",
       "          oneOf: ['+'],\n",
       "          type: 'filter'\n",
       "        })],\n",
       "        mark: 'triangleRight',\n",
       "        row: Row({\n",
       "          domain: ['+', '-'],\n",
       "          field: 'strand',\n",
       "          type: 'nominal'\n",
       "        }),\n",
       "        size: SizeValue({\n",
       "          value: 13\n",
       "        }),\n",
       "        stroke: StrokeValue({\n",
       "          value: 'white'\n",
       "        }),\n",
       "        strokeWidth: StrokeWidthValue({\n",
       "          value: 1\n",
       "        }),\n",
       "        style: Style({\n",
       "          backgroundOpacity: 0\n",
       "        }),\n",
       "        x: X({\n",
       "          field: 'start',\n",
       "          type: 'genomic'\n",
       "        })\n",
       "      }), Track({\n",
       "        color: ColorValue({\n",
       "          value: '#029F73'\n",
       "        }),\n",
       "        data: {'type': 'beddb', 'url': 'https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation', 'genomicFields': [{'index': 1, 'name': 'start'}, {'index': 2, 'name': 'end'}], 'valueFields': [{'index': 5, 'name': 'strand', 'type': 'nominal'}, {'index': 3, 'name': 'name', 'type': 'nominal'}]},\n",
       "        dataTransform: [FilterTransform({\n",
       "          field: 'strand',\n",
       "          oneOf: ['-'],\n",
       "          type: 'filter'\n",
       "        })],\n",
       "        mark: 'triangleLeft',\n",
       "        row: Row({\n",
       "          domain: ['+', '-'],\n",
       "          field: 'strand',\n",
       "          type: 'nominal'\n",
       "        }),\n",
       "        size: SizeValue({\n",
       "          value: 13\n",
       "        }),\n",
       "        stroke: StrokeValue({\n",
       "          value: 'white'\n",
       "        }),\n",
       "        strokeWidth: StrokeWidthValue({\n",
       "          value: 1\n",
       "        }),\n",
       "        style: Style({\n",
       "          backgroundOpacity: 0\n",
       "        }),\n",
       "        title: 'HFFC6_CTCF',\n",
       "        x: X({\n",
       "          field: 'start',\n",
       "          type: 'genomic'\n",
       "        })\n",
       "      })),\n",
       "      width: 25.0\n",
       "    })),\n",
       "    yOffset: 134.89208633093526\n",
       "  }), View({\n",
       "    spacing: 30,\n",
       "    views: [View({\n",
       "      arrangement: 'vertical',\n",
       "      spacing: 0,\n",
       "      views: (View({\n",
       "        alignment: 'stack',\n",
       "        tracks: (Track({\n",
       "          color: ColorValue({\n",
       "            value: 'darkgreen'\n",
       "          }),\n",
       "          data: {'type': 'bigwig', 'url': 'https://s3.amazonaws.com/gosling-lang.org/data/HFFc6_H3K4me3.bigWig', 'column': 'position', 'value': 'peak'},\n",
       "          height: 25.0,\n",
       "          mark: 'bar',\n",
       "          title: 'HFFc6_H3K4me3',\n",
       "          width: 375,\n",
       "          x: X({\n",
       "            axis: 'top',\n",
       "            field: 'start',\n",
       "            type: 'genomic'\n",
       "          }),\n",
       "          xe: Xe({\n",
       "            field: 'end',\n",
       "            type: 'genomic'\n",
       "          }),\n",
       "          y: Y({\n",
       "            field: 'peak',\n",
       "            type: 'quantitative'\n",
       "          })\n",
       "        }), Track({\n",
       "          color: ColorValue({\n",
       "            value: '#E79F00'\n",
       "          }),\n",
       "          data: {'type': 'bigwig', 'url': 'https://s3.amazonaws.com/gosling-lang.org/data/HFFc6_Atacseq.mRp.clN.bigWig', 'column': 'position', 'value': 'peak'},\n",
       "          height: 25.0,\n",
       "          mark: 'bar',\n",
       "          title: 'HFFc6_ATAC',\n",
       "          width: 375,\n",
       "          x: X({\n",
       "            field: 'start',\n",
       "            type: 'genomic'\n",
       "          }),\n",
       "          xe: Xe({\n",
       "            field: 'end',\n",
       "            type: 'genomic'\n",
       "          }),\n",
       "          y: Y({\n",
       "            axis: 'none',\n",
       "            field: 'peak',\n",
       "            type: 'quantitative'\n",
       "          })\n",
       "        }), View({\n",
       "          alignment: 'overlay',\n",
       "          height: 25.0,\n",
       "          tracks: (Track({\n",
       "            color: ColorValue({\n",
       "              value: '#0072B2'\n",
       "            }),\n",
       "            data: {'type': 'bigwig', 'url': 'https://s3.amazonaws.com/gosling-lang.org/data/HFFC6_CTCF.mRp.clN.bigWig', 'column': 'position', 'value': 'peak'},\n",
       "            mark: 'bar',\n",
       "            x: X({\n",
       "              field: 'start',\n",
       "              type: 'genomic'\n",
       "            }),\n",
       "            xe: Xe({\n",
       "              field: 'end',\n",
       "              type: 'genomic'\n",
       "            }),\n",
       "            y: Y({\n",
       "              axis: 'none',\n",
       "              field: 'peak',\n",
       "              type: 'quantitative'\n",
       "            })\n",
       "          }), Track({\n",
       "            color: ColorValue({\n",
       "              value: '#CB7AA7'\n",
       "            }),\n",
       "            data: {'type': 'beddb', 'url': 'https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation', 'genomicFields': [{'index': 1, 'name': 'start'}, {'index': 2, 'name': 'end'}], 'valueFields': [{'index': 5, 'name': 'strand', 'type': 'nominal'}, {'index': 3, 'name': 'name', 'type': 'nominal'}]},\n",
       "            dataTransform: [FilterTransform({\n",
       "              field: 'strand',\n",
       "              oneOf: ['+'],\n",
       "              type: 'filter'\n",
       "            })],\n",
       "            mark: 'triangleRight',\n",
       "            row: Row({\n",
       "              domain: ['+', '-'],\n",
       "              field: 'strand',\n",
       "              type: 'nominal'\n",
       "            }),\n",
       "            size: SizeValue({\n",
       "              value: 13\n",
       "            }),\n",
       "            stroke: StrokeValue({\n",
       "              value: 'white'\n",
       "            }),\n",
       "            strokeWidth: StrokeWidthValue({\n",
       "              value: 1\n",
       "            }),\n",
       "            style: Style({\n",
       "              backgroundOpacity: 0\n",
       "            }),\n",
       "            x: X({\n",
       "              field: 'start',\n",
       "              type: 'genomic'\n",
       "            })\n",
       "          }), Track({\n",
       "            color: ColorValue({\n",
       "              value: '#029F73'\n",
       "            }),\n",
       "            data: {'type': 'beddb', 'url': 'https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation', 'genomicFields': [{'index': 1, 'name': 'start'}, {'index': 2, 'name': 'end'}], 'valueFields': [{'index': 5, 'name': 'strand', 'type': 'nominal'}, {'index': 3, 'name': 'name', 'type': 'nominal'}]},\n",
       "            dataTransform: [FilterTransform({\n",
       "              field: 'strand',\n",
       "              oneOf: ['-'],\n",
       "              type: 'filter'\n",
       "            })],\n",
       "            mark: 'triangleLeft',\n",
       "            row: Row({\n",
       "              domain: ['+', '-'],\n",
       "              field: 'strand',\n",
       "              type: 'nominal'\n",
       "            }),\n",
       "            size: SizeValue({\n",
       "              value: 13\n",
       "            }),\n",
       "            stroke: StrokeValue({\n",
       "              value: 'white'\n",
       "            }),\n",
       "            strokeWidth: StrokeWidthValue({\n",
       "              value: 1\n",
       "            }),\n",
       "            style: Style({\n",
       "              backgroundOpacity: 0\n",
       "            }),\n",
       "            title: 'HFFC6_CTCF',\n",
       "            x: X({\n",
       "              field: 'start',\n",
       "              type: 'genomic'\n",
       "            })\n",
       "          })),\n",
       "          width: 375\n",
       "        }))\n",
       "      }), View({\n",
       "        tracks: [Track({\n",
       "          color: Color({\n",
       "            field: 'value',\n",
       "            range: 'warm',\n",
       "            type: 'quantitative'\n",
       "          }),\n",
       "          data: {'type': 'matrix', 'url': 'https://server.gosling-lang.org/api/v1/tileset_info/?d=hffc6-microc-hg38'},\n",
       "          height: 375,\n",
       "          mark: 'rect',\n",
       "          title: 'HFFc6_Micro-C',\n",
       "          width: 375,\n",
       "          x: X({\n",
       "            field: 'xs',\n",
       "            type: 'genomic'\n",
       "          }),\n",
       "          xe: Xe({\n",
       "            field: 'xe',\n",
       "            type: 'genomic'\n",
       "          }),\n",
       "          y: Y({\n",
       "            field: 'ys',\n",
       "            type: 'genomic'\n",
       "          }),\n",
       "          ye: Ye({\n",
       "            field: 'ye',\n",
       "            type: 'genomic'\n",
       "          })\n",
       "        })]\n",
       "      }), View({\n",
       "        tracks: [Track({\n",
       "          color: Color({\n",
       "            field: 'category',\n",
       "            range: ['#FF0000', '#FF4500', '#32CD32', '#008000', '#006400', '#C2E105', '#FFFF00', '#66CDAA', '#8A91D0', '#CD5C5C', '#E9967A', '#BDB76B', '#808080', '#C0C0C0', 'gray'],\n",
       "            type: 'nominal'\n",
       "          }),\n",
       "          data: {'type': 'multivec', 'url': 'https://server.gosling-lang.org/api/v1/tileset_info/?d=epilogos-hg38', 'row': 'category', 'column': 'position', 'value': 'value', 'categories': ['Active TSS', 'Flanking Active TSS', \"Transcr at gene 5\\\\' and 3\\\\'\", 'Strong transcription', 'Weak transcription', 'Genic enhancers', 'Enhancers', 'ZNF genes & repeats', 'Heterochromatin', 'Bivalent/Poised TSS', 'Flanking Bivalent TSS/Enh', 'Bivalent Enhancer', 'Repressed PolyComb', 'Weak Repressed PolyComb', 'Quiescent/Low']},\n",
       "          dataTransform: [FilterTransform({\n",
       "            field: 'value',\n",
       "            inRange: [0, 999999],\n",
       "            type: 'filter'\n",
       "          })],\n",
       "          height: 25.0,\n",
       "          mark: 'bar',\n",
       "          title: 'Epilogos (hg38)',\n",
       "          width: 375,\n",
       "          x: X({\n",
       "            axis: 'none',\n",
       "            field: 'start',\n",
       "            type: 'genomic'\n",
       "          }),\n",
       "          xe: Xe({\n",
       "            field: 'end',\n",
       "            type: 'genomic'\n",
       "          }),\n",
       "          y: Y({\n",
       "            axis: 'none',\n",
       "            field: 'value',\n",
       "            type: 'quantitative'\n",
       "          })\n",
       "        })]\n",
       "      }))\n",
       "    }), View({\n",
       "      arrangement: 'vertical',\n",
       "      spacing: 0,\n",
       "      views: (View({\n",
       "        alignment: 'stack',\n",
       "        tracks: (Track({\n",
       "          color: ColorValue({\n",
       "            value: 'darkgreen'\n",
       "          }),\n",
       "          data: {'type': 'bigwig', 'url': 'https://s3.amazonaws.com/gosling-lang.org/data/HFFc6_H3K4me3.bigWig', 'column': 'position', 'value': 'peak'},\n",
       "          height: 25.0,\n",
       "          mark: 'bar',\n",
       "          title: 'HFFc6_H3K4me3',\n",
       "          width: 375,\n",
       "          x: X({\n",
       "            axis: 'top',\n",
       "            field: 'start',\n",
       "            type: 'genomic'\n",
       "          }),\n",
       "          xe: Xe({\n",
       "            field: 'end',\n",
       "            type: 'genomic'\n",
       "          }),\n",
       "          y: Y({\n",
       "            field: 'peak',\n",
       "            type: 'quantitative'\n",
       "          })\n",
       "        }), Track({\n",
       "          color: ColorValue({\n",
       "            value: '#E79F00'\n",
       "          }),\n",
       "          data: {'type': 'bigwig', 'url': 'https://s3.amazonaws.com/gosling-lang.org/data/HFFc6_Atacseq.mRp.clN.bigWig', 'column': 'position', 'value': 'peak'},\n",
       "          height: 25.0,\n",
       "          mark: 'bar',\n",
       "          title: 'HFFc6_ATAC',\n",
       "          width: 375,\n",
       "          x: X({\n",
       "            field: 'start',\n",
       "            type: 'genomic'\n",
       "          }),\n",
       "          xe: Xe({\n",
       "            field: 'end',\n",
       "            type: 'genomic'\n",
       "          }),\n",
       "          y: Y({\n",
       "            axis: 'none',\n",
       "            field: 'peak',\n",
       "            type: 'quantitative'\n",
       "          })\n",
       "        }), View({\n",
       "          alignment: 'overlay',\n",
       "          height: 25.0,\n",
       "          tracks: (Track({\n",
       "            color: ColorValue({\n",
       "              value: '#0072B2'\n",
       "            }),\n",
       "            data: {'type': 'bigwig', 'url': 'https://s3.amazonaws.com/gosling-lang.org/data/HFFC6_CTCF.mRp.clN.bigWig', 'column': 'position', 'value': 'peak'},\n",
       "            mark: 'bar',\n",
       "            x: X({\n",
       "              field: 'start',\n",
       "              type: 'genomic'\n",
       "            }),\n",
       "            xe: Xe({\n",
       "              field: 'end',\n",
       "              type: 'genomic'\n",
       "            }),\n",
       "            y: Y({\n",
       "              axis: 'none',\n",
       "              field: 'peak',\n",
       "              type: 'quantitative'\n",
       "            })\n",
       "          }), Track({\n",
       "            color: ColorValue({\n",
       "              value: '#CB7AA7'\n",
       "            }),\n",
       "            data: {'type': 'beddb', 'url': 'https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation', 'genomicFields': [{'index': 1, 'name': 'start'}, {'index': 2, 'name': 'end'}], 'valueFields': [{'index': 5, 'name': 'strand', 'type': 'nominal'}, {'index': 3, 'name': 'name', 'type': 'nominal'}]},\n",
       "            dataTransform: [FilterTransform({\n",
       "              field: 'strand',\n",
       "              oneOf: ['+'],\n",
       "              type: 'filter'\n",
       "            })],\n",
       "            mark: 'triangleRight',\n",
       "            row: Row({\n",
       "              domain: ['+', '-'],\n",
       "              field: 'strand',\n",
       "              type: 'nominal'\n",
       "            }),\n",
       "            size: SizeValue({\n",
       "              value: 13\n",
       "            }),\n",
       "            stroke: StrokeValue({\n",
       "              value: 'white'\n",
       "            }),\n",
       "            strokeWidth: StrokeWidthValue({\n",
       "              value: 1\n",
       "            }),\n",
       "            style: Style({\n",
       "              backgroundOpacity: 0\n",
       "            }),\n",
       "            x: X({\n",
       "              field: 'start',\n",
       "              type: 'genomic'\n",
       "            })\n",
       "          }), Track({\n",
       "            color: ColorValue({\n",
       "              value: '#029F73'\n",
       "            }),\n",
       "            data: {'type': 'beddb', 'url': 'https://server.gosling-lang.org/api/v1/tileset_info/?d=gene-annotation', 'genomicFields': [{'index': 1, 'name': 'start'}, {'index': 2, 'name': 'end'}], 'valueFields': [{'index': 5, 'name': 'strand', 'type': 'nominal'}, {'index': 3, 'name': 'name', 'type': 'nominal'}]},\n",
       "            dataTransform: [FilterTransform({\n",
       "              field: 'strand',\n",
       "              oneOf: ['-'],\n",
       "              type: 'filter'\n",
       "            })],\n",
       "            mark: 'triangleLeft',\n",
       "            row: Row({\n",
       "              domain: ['+', '-'],\n",
       "              field: 'strand',\n",
       "              type: 'nominal'\n",
       "            }),\n",
       "            size: SizeValue({\n",
       "              value: 13\n",
       "            }),\n",
       "            stroke: StrokeValue({\n",
       "              value: 'white'\n",
       "            }),\n",
       "            strokeWidth: StrokeWidthValue({\n",
       "              value: 1\n",
       "            }),\n",
       "            style: Style({\n",
       "              backgroundOpacity: 0\n",
       "            }),\n",
       "            title: 'HFFC6_CTCF',\n",
       "            x: X({\n",
       "              field: 'start',\n",
       "              type: 'genomic'\n",
       "            })\n",
       "          })),\n",
       "          width: 375\n",
       "        }))\n",
       "      }), View({\n",
       "        tracks: [Track({\n",
       "          color: Color({\n",
       "            field: 'value',\n",
       "            range: 'warm',\n",
       "            type: 'quantitative'\n",
       "          }),\n",
       "          data: {'type': 'matrix', 'url': 'https://server.gosling-lang.org/api/v1/tileset_info/?d=hffc6-hic-hg38'},\n",
       "          height: 375,\n",
       "          mark: 'rect',\n",
       "          title: 'HFFc6_Hi-C',\n",
       "          width: 375,\n",
       "          x: X({\n",
       "            field: 'xs',\n",
       "            type: 'genomic'\n",
       "          }),\n",
       "          xe: Xe({\n",
       "            field: 'xe',\n",
       "            type: 'genomic'\n",
       "          }),\n",
       "          y: Y({\n",
       "            field: 'ys',\n",
       "            type: 'genomic'\n",
       "          }),\n",
       "          ye: Ye({\n",
       "            field: 'ye',\n",
       "            type: 'genomic'\n",
       "          })\n",
       "        })]\n",
       "      }), View({\n",
       "        tracks: [Track({\n",
       "          color: Color({\n",
       "            field: 'category',\n",
       "            range: ['#FF0000', '#FF4500', '#32CD32', '#008000', '#006400', '#C2E105', '#FFFF00', '#66CDAA', '#8A91D0', '#CD5C5C', '#E9967A', '#BDB76B', '#808080', '#C0C0C0', 'gray'],\n",
       "            type: 'nominal'\n",
       "          }),\n",
       "          data: {'type': 'multivec', 'url': 'https://server.gosling-lang.org/api/v1/tileset_info/?d=epilogos-hg38', 'row': 'category', 'column': 'position', 'value': 'value', 'categories': ['Active TSS', 'Flanking Active TSS', \"Transcr at gene 5\\\\' and 3\\\\'\", 'Strong transcription', 'Weak transcription', 'Genic enhancers', 'Enhancers', 'ZNF genes & repeats', 'Heterochromatin', 'Bivalent/Poised TSS', 'Flanking Bivalent TSS/Enh', 'Bivalent Enhancer', 'Repressed PolyComb', 'Weak Repressed PolyComb', 'Quiescent/Low']},\n",
       "          dataTransform: [FilterTransform({\n",
       "            field: 'value',\n",
       "            inRange: [0, 999999],\n",
       "            type: 'filter'\n",
       "          })],\n",
       "          height: 25.0,\n",
       "          mark: 'bar',\n",
       "          title: 'Epilogos (hg38)',\n",
       "          width: 375,\n",
       "          x: X({\n",
       "            axis: 'none',\n",
       "            field: 'start',\n",
       "            type: 'genomic'\n",
       "          }),\n",
       "          xe: Xe({\n",
       "            field: 'end',\n",
       "            type: 'genomic'\n",
       "          }),\n",
       "          y: Y({\n",
       "            axis: 'none',\n",
       "            field: 'value',\n",
       "            type: 'quantitative'\n",
       "          })\n",
       "        })]\n",
       "      }))\n",
       "    })]\n",
       "  })),\n",
       "  xDomain: GenomicDomain({\n",
       "    chromosome: 'chr7',\n",
       "    interval: [77700000, 81000000]\n",
       "  })\n",
       "})"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "matrix_base = gos.Track(micro_c).mark_rect().encode(\n",
    "    x=\"xs:G\",\n",
    "    xe=\"xe:G\",\n",
    "    y=\"ys:G\",\n",
    "    ye=\"ye:G\",\n",
    "    # use the values to fill in colors\n",
    "    color=gos.Color(\"value:Q\", range=\"warm\"),\n",
    ").properties(title=\"HFFc6_Micro-C\", width=375, height=375)\n",
    "\n",
    "# create bottom track\n",
    "bottom = gos.Track(epilogos).mark_bar().encode(\n",
    "    x=gos.X(\"start:G\", axis=\"none\"),\n",
    "    xe=\"end:G\",\n",
    "    y=gos.Y(\"value:Q\", axis=\"none\"),\n",
    "    # color sections (from site)\n",
    "    color=gos.Color(\"category:N\", range=[\n",
    "        \"#FF0000\", \"#FF4500\", \"#32CD32\", \"#008000\", \"#006400\",\n",
    "        \"#C2E105\", \"#FFFF00\", \"#66CDAA\", \"#8A91D0\", \"#CD5C5C\",\n",
    "        \"#E9967A\", \"#BDB76B\", \"#808080\", \"#C0C0C0\", \"gray\"\n",
    "    ]),\n",
    "    # if there are valid values\n",
    ").transform_filter(\"value\", inRange=[0, 999999]).properties(\n",
    "    title=\"Epilogos (hg38)\",\n",
    "    width=375,\n",
    "    height=375/15,\n",
    ")\n",
    "\n",
    "# make micro c mat based on vase\n",
    "left_matrix = gos.vertical(\n",
    "    top_stack, \n",
    "    matrix_base, \n",
    "    bottom, \n",
    "    # spacing at 0\n",
    "    spacing=0\n",
    ")\n",
    "\n",
    "# make right from base\n",
    "right_matrix = gos.vertical(\n",
    "    top_stack,\n",
    "    # change data to hic data\n",
    "    matrix_base.properties(\n",
    "        data=hi_c, \n",
    "        title=\"HFFc6_Hi-C\"),\n",
    "    bottom,\n",
    "    spacing=0,\n",
    ")\n",
    "\n",
    "# To combine, we align the left_stack and the matrices horizontally.\n",
    "\n",
    "# first, make combined matrix view\n",
    "combined_mat_view = gos.View(\n",
    "    views=[left_matrix, right_matrix],\n",
    "    spacing=30\n",
    ")\n",
    "\n",
    "# now, horiz combine with left_stack\n",
    "full_view=gos.horizontal(\n",
    "    left_stack,\n",
    "    combined_mat_view\n",
    ").properties(\n",
    "    title=\"Matrix Visualization Recreation\",\n",
    "    # add a zoom \n",
    "    xDomain=gos.GenomicDomain(chromosome=\"chr7\", interval=[77700000, 81000000]),\n",
    "    spacing=1,\n",
    "    linkingId=\"-\"\n",
    ")\n",
    "\n",
    "full_view.save(\"visuals/combined-Hic/combined-Hic-MicroC-Maps.json\")\n",
    "display(full_view)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "hidive-env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
